{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<!---\n",
    "Latex Macros\n",
    "-->\n",
    "$$\n",
    "\\newcommand{\\bar}{\\,|\\,}\n",
    "\\newcommand{\\Xs}{\\mathcal{X}}\n",
    "\\newcommand{\\Ys}{\\mathcal{Y}}\n",
    "\\newcommand{\\y}{\\mathbf{y}}\n",
    "\\newcommand{\\weights}{\\mathbf{w}}\n",
    "\\newcommand{\\balpha}{\\boldsymbol{\\alpha}}\n",
    "\\newcommand{\\bbeta}{\\boldsymbol{\\beta}}\n",
    "\\newcommand{\\aligns}{\\mathbf{a}}\n",
    "\\newcommand{\\align}{a}\n",
    "\\newcommand{\\source}{\\mathbf{s}}\n",
    "\\newcommand{\\target}{\\mathbf{t}}\n",
    "\\newcommand{\\ssource}{s}\n",
    "\\newcommand{\\starget}{t}\n",
    "\\newcommand{\\repr}{\\mathbf{f}}\n",
    "\\newcommand{\\repry}{\\mathbf{g}}\n",
    "\\newcommand{\\x}{\\mathbf{x}}\n",
    "\\newcommand{\\prob}{p}\n",
    "\\newcommand{\\vocab}{V}\n",
    "\\newcommand{\\params}{\\boldsymbol{\\theta}}\n",
    "\\newcommand{\\param}{\\theta}\n",
    "\\DeclareMathOperator{\\perplexity}{PP}\n",
    "\\DeclareMathOperator{\\argmax}{argmax}\n",
    "\\DeclareMathOperator{\\argmin}{argmin}\n",
    "\\newcommand{\\train}{\\mathcal{D}}\n",
    "\\newcommand{\\counts}[2]{\\#_{#1}(#2) }\n",
    "\\newcommand{\\length}[1]{\\text{length}(#1) }\n",
    "\\newcommand{\\indi}{\\mathbb{I}}\n",
    "$$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Assignment 2\n",
    "\n",
    "## Introduction\n",
    "In this assignment you will build the first stage of a biomedical event extractor. Biomedical events are state changes of biomolecules. For example, if you have a protein and you add a phosphate (PO4) group to it, this is referred to as a phosphorylation event. Many papers in the biomedical literature mention such events. The grand goal of biomedical event extraction is to teach machines how to read this literature and produce structured representations of biomedical events that biomedical researchers can query effectively. This task has received considerable attention in the NLP literature, and is the topic of a biennial [shared task](http://2011.bionlp-st.org/). We will use the data from this task as starting point for this assignment.   \n",
    "\n",
    "To illustrate biomedical event extraction, let us consider an example. From the sentence \n",
    "\n",
    "> **phosphorylation** of TRAF2 **inhibits** **binding** to the CD40 domain\n",
    "\n",
    "we could extract the structure \n",
    "\n",
    "> Negative_Regulation(Phosphorylation(TRAF2), Binding(TRAF2, CD40)\n",
    "\n",
    "and store it in a database. Someone can then query this database, for example, to figure out all ways to prevent binding of TRAF2 to CD40.\n",
    "\n",
    "The task is often divided into two steps. First you need to find **trigger** words in the sentence that correspond to biomedical events, and determine their event type *label*. For example, in the above sentence \"phosphorylation\" is a trigger word for an event of type \"Phosphorylation\", \"inhibits\" a trigger word for a \"Negative Regulation\" event, and \"binding\" a trigger for a \"Binding\" event. Notice that sometimes the type labels are obvious, but often they are not. Also note that the label of a word could be \"None\". For example, the word \"of\" in the above sentence has the label \"None\".  \n",
    "\n",
    "The second step requires the extractor to produce **argument relations** between event triggers and protein mentions or other event triggers. For example, in the above case the argument of \"phosphorylation\" is \"TRAF2\", and one argument of \"inhibits\" is \"phosphorylation of TRAF2\" whereas the other is \"binding to the CD40 domain\". In this assignment you **do not have to do this**. We will focus on the event trigger detection problem exclusively. \n",
    "\n",
    "## Goal\n",
    "Your goal is to develop an event trigger labeler. This extractor is given a sentence and a candidate token. Both constitute the input $\\x$. One such input could be: \n",
    "\n",
    "> $\\x$: phosphorylation of TRAF2 **inhibits** binding to the CD40 domain\n",
    "\n",
    "The goal is to predict the label $y$ of the candidate event trigger. In the above case the label would be $y=\\text{Negative_Regulation}$. \n",
    "\n",
    "Some candidates may not refer to event triggers at all. For example:\n",
    "\n",
    "> $\\x$: phosphorylation **of** TRAF2 inhibits binding to the CD40 domain\n",
    "\n",
    "In such cases the label is $y=\\text{None}$.\n",
    "\n",
    "## Resources\n",
    "To develop your model you have access to:\n",
    "\n",
    "* The data in `data/bionlp/train`. This data can be split into training and dev set (as done below), or used for cross-validation.\n",
    "* Helper code stored in the python module [bio.py](/edit/statnlpbook/bio.py).\n",
    "* Libraries on the [docker image](https://github.com/uclmr/stat-nlp-book/blob/python/Dockerfile) which contains everything in [this image](https://github.com/jupyter/docker-stacks/tree/master/scipy-notebook), including scikit-learn and tensorflow. \n",
    "\n",
    "As we have to run the notebooks of all students, and because writing efficient code is important, **your notebook should run in 5 minutes at most**, on your machine. Further comments:\n",
    "\n",
    "* We have tested a possible solution on the Azure VMs and it ran in about 30s, so it is possible to train a reasonable model on the data in reasonable time. If you find training times too long for your development cycle you can reduce the training set size. Once you have found a good solution you can increase the size again. Caveat: model parameters tuned on a smaller dataset may not be optimal for a larger training set.\n",
    "\n",
    "* Try to run your parameter optimisation offline, such that in your answer notebook the best parameters are already set and don't need to be searched. Include your optimisation code in the notebook, but don't call it at each notebook run."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Hint\n",
    "While you do not need to predict the arguments of an event, it is important to understand how trigger labels relate to the syntactic and semantic arguments of the trigger word. Features that can capture this relation might help you in improving the result. Do inspect the data and try to get an understanding of it. That said, you don't have to be a biomedical expert to do well in this task. A few of the best results on the task were achieved by NLP researchers without any biomedical experience. They would, however, still inspect the data carefully.  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup Instructions\n",
    "It is important that this file is placed in the **correct directory**. It will not run otherwise. The correct directory is\n",
    "\n",
    "    DIRECTORY_OF_YOUR_BOOK/assignments/2017/assignment2/problem/\n",
    "    \n",
    "where `DIRECTORY_OF_YOUR_BOOK` is a placeholder for the directory you downloaded the book to. After you placed it there, **rename the file** to your UCL ID (of the form `ucxxxxx`). "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## General Instructions\n",
    "This notebook will be used by you to provide your solution, and by us to both assess your solution and enter your marks. It contains three types of sections:\n",
    "\n",
    "1. **Setup** Sections: these sections set up code and resources for assessment. **Do not edit, move nor copy these cells**.\n",
    "2. **Assessment** Sections: these sections are used for both evaluating the output of your code, and for markers to enter their marks. **Do not edit, move, nor copy these cells**.\n",
    "3. **Task** Sections: these sections require your solutions. They may contain stub code, and you are expected to edit this code. For free text answers simply edit the markdown field.  \n",
    "\n",
    "**If you edit, move or copy any of the setup, assessments and mark cells, you will be penalised with -10 points**.\n",
    "\n",
    "Note that you are free to **create additional notebook cells** within a task section. \n",
    "\n",
    "Please **do not share** this assignment publicly, by uploading it online, emailing it to friends etc. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Submission Instructions\n",
    "\n",
    "To submit your solution:\n",
    "\n",
    "* Make sure that your solution is fully contained in this notebook. \n",
    "* Make sure that your solution runs linearly from start to end (no execution hops). We will run your notebook in that order.\n",
    "* **If running your notebook produces a trivially fixable error that we spot, we will correct it and penalise you with -10 points. Otherwise you will get 0 points for that solution.**\n",
    "* **Rename this notebook to your UCL ID** (of the form \"ucxxxxx\"), if you have not already done so. ** Failure to do so will result in -1 point.**\n",
    "* Download the notebook in Jupyter via *File -> Download as -> Notebook (.ipynb)*.\n",
    "* Upload the notebook to the Moodle submission site."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <font color='green'>Setup 1</font>: Load Libraries\n",
    "This cell loads libraries important for evaluation and assessment of your model. **Do not change, move or copy it.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%%capture\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%matplotlib inline\n",
    "#! SETUP 1 - DO NOT CHANGE, MOVE NOR COPY\n",
    "import sys, os\n",
    "_snlp_book_dir = \"../../../../\"\n",
    "sys.path.append(_snlp_book_dir) \n",
    "import math\n",
    "from collections import defaultdict\n",
    "import statnlpbook.bio as bio"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <font color='green'>Setup 2</font>: Load Training Data\n",
    "\n",
    "This cell loads the training data. **Do not edit this setup section, nor copy it**. Instead refer to the variables in your own code, and slice and dice them as you see fit (but do not change their values). For example, no one will stop you from introducing, in the corresponding task section, `my_event_train` and `my_event_dev` variables that split the data into different folds.   \n",
    "\n",
    "Notice that the data is loaded from `json` files like [this one](/edit/data/bionlp/train/PMC-1310901-00-TIAB.json). Generally, you do not need to understand this format, as we provide loading functions that produce more convenient data structures shown below. But do feel free to investigate. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#! SETUP 2 - DO NOT CHANGE, MOVE NOR COPY\n",
    "train_path = _snlp_book_dir + \"data/bionlp/train\"\n",
    "event_corpus = bio.load_assignment2_training_data(train_path)\n",
    "event_train = event_corpus[:len(event_corpus)//4 * 3]\n",
    "event_dev = event_corpus[len(event_corpus)//4 * 3:]\n",
    "assert(len(event_train)==53988)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Structures\n",
    "The data comes in the form of pairs consisting of `EventCandidate` objects and their trigger labels. The `EventCandidate` class can be found in [bio.py](/edit/statnlpbook/bio.py)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(<statnlpbook.bio.EventCandidate at 0x7f1b26c0eba8>, 'None')"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "event_candidate, label = event_corpus[90]\n",
    "(event_candidate, label)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Event candidate objects specify the classification problem. They consist of a sentence `sent` and the position `trigger_index` of the trigger candidate word. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "INTRODUCTION Chronic myeloid leukemia ( CML ) is a clonal myeloproliferative disorder with a typical three phased course ( chronic , accelerated and blastic phase ) reflecting the loss of differentiation and malignant progress which inevitably leads to death after the blastic phase ( 1,2 ) ."
      ],
      "text/plain": [
       "<statnlpbook.bio.Sentence at 0x7f1b26b46f28>"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "event_candidate.sent\n",
    "#event_candidate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "36"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "event_candidate.trigger_index"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Event candidates also have a set of candidate arguments. These point to token spans (index of first token, inclusive, index of last token, exclusive) in the sentence that may or may not be *arguments* of the event. In the full event extraction task one needs to predict which of these candidates are true arguments of the events. However, here we will ignore this task, and give you only the information what candidates exist, not what their labels are. Note that this information can still be **very important** to understand what type of event the candidate corresponds to, if any. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(2, 6), (23, 24), (24, 25), (59, 63)]"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "event_candidate, label = event_corpus[11]\n",
    "event_candidate.argument_candidate_spans[:4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "First , treatment <font color='red'>[</font>of<font color='red'>]</font> <font color='red'>[</font><font color='blue'>[IRF-4]</font><font color='red'>]</font> <font color='red'>[</font>-negative<font color='red'>]</font> lymphoid , myeloid and monocytic cell lines <font color='red'>[</font>with<font color='red'>]</font> the <font color='red'>[</font>methylation-inhibitor<font color='red'>]</font> 5-aza-2-deoxycytidine <font color='green'>resulted</font> in a time <font color='red'>[</font>-<font color='red'>]</font> and <font color='red'>[</font>concentration-dependent<font color='red'>]</font> <font color='red'>[</font>increase<font color='red'>]</font> <font color='red'>[</font>of<font color='red'>]</font> <font color='red'>[</font><font color='blue'>[IRF-4]</font><font color='red'>]</font> <font color='red'>[</font>mRNA<font color='red'>]</font> and <font color='red'>[</font>protein<font color='red'>]</font> <font color='red'>[</font>levels<font color='red'>]</font> ."
      ],
      "text/plain": [
       "<statnlpbook.bio.EventCandidate at 0x7fb01b190a20>"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "event_candidate, label = event_corpus[49]\n",
    "event_candidate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "    <div id='displacy27' style=\"overflow: scroll; width: 5000px;\"></div>\n",
       "    <script>\n",
       "    $(function() {\n",
       "    requirejs.config({\n",
       "        paths: {\n",
       "            'displaCy': ['/files/node_modules/displacy/displacy'],\n",
       "                                                  // strip .js ^, require adds it back\n",
       "        },\n",
       "    });\n",
       "    require(['displaCy'], function() {\n",
       "        console.log(\"Loaded :)\");\n",
       "        const displacy = new displaCy('http://localhost:8000', {\n",
       "            container: '#displacy27',\n",
       "            format: 'spacy',\n",
       "            distance: 150,\n",
       "            offsetX: 0,\n",
       "            wordSpacing: 20,\n",
       "            arrowSpacing: 3,\n",
       "\n",
       "        });\n",
       "        const parse = {\n",
       "            arcs: [{\"dir\": \"left\", \"start\": 0, \"label\": \"advmod\", \"end\": 17}, {\"dir\": \"left\", \"start\": 2, \"label\": \"nsubj\", \"end\": 17}, {\"dir\": \"left\", \"start\": 4, \"label\": \"nn\", \"end\": 12}, {\"dir\": \"left\", \"start\": 5, \"label\": \"amod\", \"end\": 12}, {\"dir\": \"left\", \"start\": 6, \"label\": \"amod\", \"end\": 12}, {\"dir\": \"right\", \"start\": 6, \"label\": \"conj_and\", \"end\": 8}, {\"dir\": \"left\", \"start\": 8, \"label\": \"amod\", \"end\": 12}, {\"dir\": \"right\", \"start\": 6, \"label\": \"conj_and\", \"end\": 10}, {\"dir\": \"left\", \"start\": 10, \"label\": \"amod\", \"end\": 12}, {\"dir\": \"left\", \"start\": 11, \"label\": \"nn\", \"end\": 12}, {\"dir\": \"right\", \"start\": 2, \"label\": \"prep_of\", \"end\": 12}, {\"dir\": \"left\", \"start\": 14, \"label\": \"det\", \"end\": 16}, {\"dir\": \"left\", \"start\": 15, \"label\": \"amod\", \"end\": 16}, {\"dir\": \"right\", \"start\": 2, \"label\": \"prep_with\", \"end\": 16}, {\"dir\": \"left\", \"start\": 19, \"label\": \"det\", \"end\": 20}, {\"dir\": \"right\", \"start\": 17, \"label\": \"prep_in\", \"end\": 20}, {\"dir\": \"left\", \"start\": 23, \"label\": \"amod\", \"end\": 24}, {\"dir\": \"right\", \"start\": 17, \"label\": \"prep_in\", \"end\": 24}, {\"dir\": \"right\", \"start\": 20, \"label\": \"conj_and\", \"end\": 24}, {\"dir\": \"left\", \"start\": 26, \"label\": \"nn\", \"end\": 27}, {\"dir\": \"right\", \"start\": 24, \"label\": \"prep_of\", \"end\": 27}, {\"dir\": \"left\", \"start\": 29, \"label\": \"nn\", \"end\": 30}, {\"dir\": \"right\", \"start\": 24, \"label\": \"prep_of\", \"end\": 30}, {\"dir\": \"right\", \"start\": 27, \"label\": \"conj_and\", \"end\": 30}],\n",
       "            words: [{\"text\": \"First\", \"tag\": \"RB\"}, {\"text\": \",\", \"tag\": \",\"}, {\"text\": \"treatment\", \"tag\": \"NN\"}, {\"text\": \"of\", \"tag\": \"IN\"}, {\"text\": \"IRF-4\", \"tag\": \"NN\"}, {\"text\": \"-negative\", \"tag\": \"JJ\"}, {\"text\": \"lymphoid\", \"tag\": \"JJ\"}, {\"text\": \",\", \"tag\": \",\"}, {\"text\": \"myeloid\", \"tag\": \"JJ\"}, {\"text\": \"and\", \"tag\": \"CC\"}, {\"text\": \"monocytic\", \"tag\": \"JJ\"}, {\"text\": \"cell\", \"tag\": \"NN\"}, {\"text\": \"lines\", \"tag\": \"NNS\"}, {\"text\": \"with\", \"tag\": \"IN\"}, {\"text\": \"the\", \"tag\": \"DT\"}, {\"text\": \"methylation-inhibitor\", \"tag\": \"JJ\"}, {\"text\": \"5-aza-2-deoxycytidine\", \"tag\": \"NN\"}, {\"text\": \"resulted\", \"tag\": \"VBD\"}, {\"text\": \"in\", \"tag\": \"IN\"}, {\"text\": \"a\", \"tag\": \"DT\"}, {\"text\": \"time\", \"tag\": \"NN\"}, {\"text\": \"-\", \"tag\": \":\"}, {\"text\": \"and\", \"tag\": \"CC\"}, {\"text\": \"concentration-dependent\", \"tag\": \"JJ\"}, {\"text\": \"increase\", \"tag\": \"NN\"}, {\"text\": \"of\", \"tag\": \"IN\"}, {\"text\": \"IRF-4\", \"tag\": \"NN\"}, {\"text\": \"mRNA\", \"tag\": \"NN\"}, {\"text\": \"and\", \"tag\": \"CC\"}, {\"text\": \"protein\", \"tag\": \"NN\"}, {\"text\": \"levels\", \"tag\": \"NNS\"}, {\"text\": \".\", \"tag\": \".\"}]\n",
       "        };\n",
       "\n",
       "        displacy.render(parse, {\n",
       "            uniqueId: 'render_displacy27'\n",
       "            //color: '#ff0000'\n",
       "        });\n",
       "        return {};\n",
       "    });\n",
       "    });\n",
       "    </script>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bio.render_dependencies(event_candidate.sent)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'begin': 4, 'end': 5, 'label': 'Protein'},\n",
       " {'begin': 26, 'end': 27, 'label': 'Protein'}]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "event_candidate.sent.mentions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can compactly visualise the complete candidate using `bio.render_event`, as shown below. Here the green span corresponds to the token at the trigger index. The spans in red brackets correspond to the argument candidates. The blue spans are protein mentions. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sentences\n",
    "The sentence object of an event candidate provides additional information about the sentence, such as what spans are proteins, what Part-of-Speech labels the tokens have, and a dependency parse of the sentence. First, the `tokens` field of a sentence provides useful features of tokens: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'begin': 873,\n",
       " 'end': 883,\n",
       " 'index': 15,\n",
       " 'pos': 'NN',\n",
       " 'stem': 'express',\n",
       " 'word': 'expression'}"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "event_candidate.sent.tokens[event_candidate.trigger_index]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `dependencies` field stores lexical dependencies between words:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'head': 44, 'label': 'nsubj', 'mod': 0}"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "event_candidate.sent.dependencies[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can render the dependency graph of a sentence like so:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "    <div id='displacy0' style=\"overflow: scroll; width: 5000px;\"></div>\n",
       "    <script>\n",
       "    $(function() {\n",
       "    requirejs.config({\n",
       "        paths: {\n",
       "            'displaCy': ['/files/node_modules/displacy/displacy'],\n",
       "                                                  // strip .js ^, require adds it back\n",
       "        },\n",
       "    });\n",
       "    require(['displaCy'], function() {\n",
       "        console.log(\"Loaded :)\");\n",
       "        const displacy = new displaCy('http://localhost:8000', {\n",
       "            container: '#displacy0',\n",
       "            format: 'spacy',\n",
       "            distance: 150,\n",
       "            offsetX: 0,\n",
       "            wordSpacing: 20,\n",
       "            arrowSpacing: 3,\n",
       "\n",
       "        });\n",
       "        const parse = {\n",
       "            arcs: [{\"dir\": \"left\", \"start\": 0, \"label\": \"nsubj\", \"end\": 44}, {\"dir\": \"left\", \"start\": 2, \"label\": \"nn\", \"end\": 4}, {\"dir\": \"left\", \"start\": 3, \"label\": \"amod\", \"end\": 4}, {\"dir\": \"left\", \"start\": 4, \"label\": \"nn\", \"end\": 7}, {\"dir\": \"right\", \"start\": 4, \"label\": \"num\", \"end\": 5}, {\"dir\": \"left\", \"start\": 6, \"label\": \"nn\", \"end\": 7}, {\"dir\": \"right\", \"start\": 0, \"label\": \"prep_of\", \"end\": 7}, {\"dir\": \"left\", \"start\": 9, \"label\": \"amod\", \"end\": 10}, {\"dir\": \"right\", \"start\": 7, \"label\": \"prep_in\", \"end\": 10}, {\"dir\": \"right\", \"start\": 7, \"label\": \"prep_due_to\", \"end\": 13}, {\"dir\": \"left\", \"start\": 15, \"label\": \"nn\", \"end\": 16}, {\"dir\": \"right\", \"start\": 13, \"label\": \"prep_of\", \"end\": 16}, {\"dir\": \"left\", \"start\": 18, \"label\": \"det\", \"end\": 20}, {\"dir\": \"left\", \"start\": 19, \"label\": \"nn\", \"end\": 20}, {\"dir\": \"right\", \"start\": 0, \"label\": \"prep_in\", \"end\": 20}, {\"dir\": \"left\", \"start\": 21, \"label\": \"mark\", \"end\": 28}, {\"dir\": \"left\", \"start\": 22, \"label\": \"det\", \"end\": 25}, {\"dir\": \"left\", \"start\": 23, \"label\": \"nn\", \"end\": 25}, {\"dir\": \"left\", \"start\": 24, \"label\": \"nn\", \"end\": 25}, {\"dir\": \"left\", \"start\": 25, \"label\": \"nsubjpass\", \"end\": 28}, {\"dir\": \"left\", \"start\": 25, \"label\": \"xsubj\", \"end\": 34}, {\"dir\": \"left\", \"start\": 26, \"label\": \"aux\", \"end\": 28}, {\"dir\": \"left\", \"start\": 27, \"label\": \"auxpass\", \"end\": 28}, {\"dir\": \"left\", \"start\": 28, \"label\": \"advcl\", \"end\": 44}, {\"dir\": \"left\", \"start\": 29, \"label\": \"aux\", \"end\": 34}, {\"dir\": \"left\", \"start\": 30, \"label\": \"cop\", \"end\": 34}, {\"dir\": \"left\", \"start\": 31, \"label\": \"det\", \"end\": 34}, {\"dir\": \"left\", \"start\": 32, \"label\": \"amod\", \"end\": 34}, {\"dir\": \"left\", \"start\": 33, \"label\": \"amod\", \"end\": 34}, {\"dir\": \"right\", \"start\": 28, \"label\": \"xcomp\", \"end\": 34}, {\"dir\": \"left\", \"start\": 36, \"label\": \"amod\", \"end\": 38}, {\"dir\": \"left\", \"start\": 37, \"label\": \"amod\", \"end\": 38}, {\"dir\": \"right\", \"start\": 34, \"label\": \"prep_in\", \"end\": 38}, {\"dir\": \"right\", \"start\": 38, \"label\": \"abbrev\", \"end\": 40}, {\"dir\": \"left\", \"start\": 43, \"label\": \"expl\", \"end\": 44}, {\"dir\": \"left\", \"start\": 45, \"label\": \"amod\", \"end\": 46}, {\"dir\": \"right\", \"start\": 44, \"label\": \"nsubj\", \"end\": 46}, {\"dir\": \"left\", \"start\": 47, \"label\": \"complm\", \"end\": 69}, {\"dir\": \"left\", \"start\": 48, \"label\": \"det\", \"end\": 49}, {\"dir\": \"left\", \"start\": 49, \"label\": \"nsubjpass\", \"end\": 69}, {\"dir\": \"left\", \"start\": 51, \"label\": \"amod\", \"end\": 52}, {\"dir\": \"right\", \"start\": 49, \"label\": \"prep_of\", \"end\": 52}, {\"dir\": \"left\", \"start\": 56, \"label\": \"det\", \"end\": 61}, {\"dir\": \"left\", \"start\": 57, \"label\": \"nn\", \"end\": 61}, {\"dir\": \"left\", \"start\": 58, \"label\": \"nn\", \"end\": 61}, {\"dir\": \"left\", \"start\": 59, \"label\": \"nn\", \"end\": 61}, {\"dir\": \"left\", \"start\": 60, \"label\": \"amod\", \"end\": 61}, {\"dir\": \"right\", \"start\": 52, \"label\": \"prep_such_as\", \"end\": 61}, {\"dir\": \"right\", \"start\": 61, \"label\": \"num\", \"end\": 62}, {\"dir\": \"right\", \"start\": 61, \"label\": \"abbrev\", \"end\": 64}, {\"dir\": \"left\", \"start\": 67, \"label\": \"auxpass\", \"end\": 69}, {\"dir\": \"left\", \"start\": 68, \"label\": \"advmod\", \"end\": 69}, {\"dir\": \"right\", \"start\": 46, \"label\": \"ccomp\", \"end\": 69}, {\"dir\": \"left\", \"start\": 71, \"label\": \"det\", \"end\": 72}, {\"dir\": \"right\", \"start\": 69, \"label\": \"prep_in\", \"end\": 72}, {\"dir\": \"right\", \"start\": 72, \"label\": \"prep_of\", \"end\": 74}],\n",
       "            words: [{\"text\": \"Down-regulation\", \"tag\": \"NN\"}, {\"text\": \"of\", \"tag\": \"IN\"}, {\"text\": \"interferon\", \"tag\": \"NN\"}, {\"text\": \"regulatory\", \"tag\": \"JJ\"}, {\"text\": \"factor\", \"tag\": \"NN\"}, {\"text\": \"4\", \"tag\": \"CD\"}, {\"text\": \"gene\", \"tag\": \"NN\"}, {\"text\": \"expression\", \"tag\": \"NN\"}, {\"text\": \"in\", \"tag\": \"IN\"}, {\"text\": \"leukemic\", \"tag\": \"JJ\"}, {\"text\": \"cells\", \"tag\": \"NNS\"}, {\"text\": \"due\", \"tag\": \"IN\"}, {\"text\": \"to\", \"tag\": \"TO\"}, {\"text\": \"hypermethylation\", \"tag\": \"NN\"}, {\"text\": \"of\", \"tag\": \"IN\"}, {\"text\": \"CpG\", \"tag\": \"NN\"}, {\"text\": \"motifs\", \"tag\": \"NNS\"}, {\"text\": \"in\", \"tag\": \"IN\"}, {\"text\": \"the\", \"tag\": \"DT\"}, {\"text\": \"promoter\", \"tag\": \"NN\"}, {\"text\": \"region\", \"tag\": \"NN\"}, {\"text\": \"Although\", \"tag\": \"IN\"}, {\"text\": \"the\", \"tag\": \"DT\"}, {\"text\": \"bcr\", \"tag\": \"NN\"}, {\"text\": \"-abl\", \"tag\": \"NN\"}, {\"text\": \"translocation\", \"tag\": \"NN\"}, {\"text\": \"has\", \"tag\": \"VBZ\"}, {\"text\": \"been\", \"tag\": \"VBN\"}, {\"text\": \"shown\", \"tag\": \"VBN\"}, {\"text\": \"to\", \"tag\": \"TO\"}, {\"text\": \"be\", \"tag\": \"VB\"}, {\"text\": \"the\", \"tag\": \"DT\"}, {\"text\": \"causative\", \"tag\": \"JJ\"}, {\"text\": \"genetic\", \"tag\": \"JJ\"}, {\"text\": \"aberration\", \"tag\": \"NN\"}, {\"text\": \"in\", \"tag\": \"IN\"}, {\"text\": \"chronic\", \"tag\": \"JJ\"}, {\"text\": \"myeloid\", \"tag\": \"JJ\"}, {\"text\": \"leukemia\", \"tag\": \"NN\"}, {\"text\": \"(\", \"tag\": \"-LRB-\"}, {\"text\": \"CML\", \"tag\": \"NN\"}, {\"text\": \")\", \"tag\": \"-RRB-\"}, {\"text\": \",\", \"tag\": \",\"}, {\"text\": \"there\", \"tag\": \"EX\"}, {\"text\": \"is\", \"tag\": \"VBZ\"}, {\"text\": \"mounting\", \"tag\": \"JJ\"}, {\"text\": \"evidence\", \"tag\": \"NN\"}, {\"text\": \"that\", \"tag\": \"IN\"}, {\"text\": \"the\", \"tag\": \"DT\"}, {\"text\": \"deregulation\", \"tag\": \"NN\"}, {\"text\": \"of\", \"tag\": \"IN\"}, {\"text\": \"other\", \"tag\": \"JJ\"}, {\"text\": \"genes\", \"tag\": \"NNS\"}, {\"text\": \",\", \"tag\": \",\"}, {\"text\": \"such\", \"tag\": \"JJ\"}, {\"text\": \"as\", \"tag\": \"IN\"}, {\"text\": \"the\", \"tag\": \"DT\"}, {\"text\": \"transcription\", \"tag\": \"NN\"}, {\"text\": \"factor\", \"tag\": \"NN\"}, {\"text\": \"interferon\", \"tag\": \"NN\"}, {\"text\": \"regulatory\", \"tag\": \"JJ\"}, {\"text\": \"factor\", \"tag\": \"NN\"}, {\"text\": \"4\", \"tag\": \"CD\"}, {\"text\": \"(\", \"tag\": \"-LRB-\"}, {\"text\": \"IRF-4\", \"tag\": \"NN\"}, {\"text\": \")\", \"tag\": \"-RRB-\"}, {\"text\": \",\", \"tag\": \",\"}, {\"text\": \"is\", \"tag\": \"VBZ\"}, {\"text\": \"also\", \"tag\": \"RB\"}, {\"text\": \"implicated\", \"tag\": \"VBN\"}, {\"text\": \"in\", \"tag\": \"IN\"}, {\"text\": \"the\", \"tag\": \"DT\"}, {\"text\": \"pathogenesis\", \"tag\": \"NN\"}, {\"text\": \"of\", \"tag\": \"IN\"}, {\"text\": \"CML\", \"tag\": \"NN\"}, {\"text\": \".\", \"tag\": \".\"}]\n",
       "        };\n",
       "\n",
       "        displacy.render(parse, {\n",
       "            uniqueId: 'render_displacy0'\n",
       "            //color: '#ff0000'\n",
       "        });\n",
       "        return {};\n",
       "    });\n",
       "    });\n",
       "    </script>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "event_candidate, label = event_corpus[2]\n",
    "bio.render_dependencies(event_candidate.sent)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<font color='red'>[</font>Down-regulation<font color='red'>]</font> <font color='red'>[</font>of<font color='red'>]</font> <font color='red'>[</font><font color='blue'>[interferon regulatory factor 4]</font><font color='red'>]</font> gene <font color='green'>expression</font> in leukemic cells <font color='red'>[</font>due<font color='red'>]</font> <font color='red'>[</font>to<font color='red'>]</font> hypermethylation <font color='red'>[</font>of<font color='red'>]</font> CpG motifs in the <font color='red'>[</font>promoter<font color='red'>]</font> region Although the <font color='red'>[</font><font color='blue'>[bcr]</font><font color='red'>]</font> <font color='red'>[</font><font color='blue'>[-abl]</font><font color='red'>]</font> <font color='red'>[</font>translocation<font color='red'>]</font> has been shown <font color='red'>[</font>to<font color='red'>]</font> be the causative genetic <font color='red'>[</font>aberration<font color='red'>]</font> in chronic myeloid leukemia ( CML ) , there is mounting evidence <font color='red'>[</font>that<font color='red'>]</font> the <font color='red'>[</font>deregulation<font color='red'>]</font> <font color='red'>[</font>of<font color='red'>]</font> other genes , such as the <font color='red'>[</font>transcription<font color='red'>]</font> <font color='red'>[</font>factor<font color='red'>]</font> <font color='red'>[</font><font color='blue'>[interferon regulatory factor 4]</font><font color='red'>]</font> ( <font color='red'>[</font><font color='blue'>[IRF-4]</font><font color='red'>]</font> ) , is also <font color='red'>[</font>implicated<font color='red'>]</font> in the pathogenesis <font color='red'>[</font>of<font color='red'>]</font> CML ."
      ],
      "text/plain": [
       "<statnlpbook.bio.EventCandidate at 0x7f4555b5a470>"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "event_candidate, label = event_corpus[2]\n",
    "#0,2,11,32,34,38,39,43,44,46,49,52,62\n",
    "print(event_candidate.trigger_index)\n",
    "#print(event_candidate.sent.mentions)\n",
    "#print(event_candidate.sent.parents,event_candidate.sent.children)\n",
    "event_candidate"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can learn about the dependency labels in the [Stanford typed dependencies manual](http://nlp.stanford.edu/software/dependencies_manual.pdf). We also provide [lecture notes on dependency parsing](/notebooks/chapters/Transition-based%20dependency%20parsing.ipynb), including various pointers to more information. \n",
    "\n",
    "The `mentions` stores which spans correspond to proteins."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'begin': 2, 'end': 6, 'label': 'Protein'},\n",
       " {'begin': 23, 'end': 24, 'label': 'Protein'},\n",
       " {'begin': 24, 'end': 25, 'label': 'Protein'},\n",
       " {'begin': 59, 'end': 63, 'label': 'Protein'},\n",
       " {'begin': 64, 'end': 65, 'label': 'Protein'}]"
      ]
     },
     "execution_count": 187,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "event_candidate, label = event_corpus[1]\n",
    "event_candidate.sent.mentions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are some convenience functions for the sentence to check all the syntactic parents or children of a token, or whether a specific token is within a protein mention. These can be useful when designing features:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "([(44, 'nsubj')], [(7, 'prep_of'), (20, 'prep_in')])"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "event_candidate.sent.parents[0], event_candidate.sent.children[0] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 296,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(True, False)"
      ]
     },
     "execution_count": 296,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "event_candidate.sent.is_protein[14], event_candidate.sent.is_protein[7]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Labels\n",
    "It is useful to know the complete set of event labels:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 239,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Binding',\n",
       " 'Gene_expression',\n",
       " 'Localization',\n",
       " 'Negative_regulation',\n",
       " 'None',\n",
       " 'Phosphorylation',\n",
       " 'Positive_regulation',\n",
       " 'Protein_catabolism',\n",
       " 'Regulation',\n",
       " 'Transcription'}"
      ]
     },
     "execution_count": 239,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "{y for _,y in event_corpus}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <font color='blue'>Task 1</font>: Create a Feature Function\n",
    "\n",
    "In this task you will extract a specific feature representation $\\repr(\\x)$ for an event candidate $\\x$. In particular, we want to add as features the syntactic children (modifiers) of the trigger token, together with their syntactic dependency label. A modifier of a token $h$ is a token $m$ that modifies $h$'s meaning. For example, in the noun phrase \"green light\" the adjective \"green\" modifies the noun \"light\". We will refer to the modifier token as the \"child\", and the modified token as \"parent\". Correspondingly, in the dependency graph modifiers are the child nodes of the modified tokens.  \n",
    "\n",
    "The feature function will have to be implemented as a python function that populates a python dictionary with key-value pairs where the key indicates both the word and syntactic label of the child. \n",
    "\n",
    "For example, consider the following event and dependency parse:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "    <div id='displacy4' style=\"overflow: scroll; width: 5000px;\"></div>\n",
       "    <script>\n",
       "    $(function() {\n",
       "    requirejs.config({\n",
       "        paths: {\n",
       "            'displaCy': ['/files/node_modules/displacy/displacy'],\n",
       "                                                  // strip .js ^, require adds it back\n",
       "        },\n",
       "    });\n",
       "    require(['displaCy'], function() {\n",
       "        console.log(\"Loaded :)\");\n",
       "        const displacy = new displaCy('http://localhost:8000', {\n",
       "            container: '#displacy4',\n",
       "            format: 'spacy',\n",
       "            distance: 150,\n",
       "            offsetX: 0,\n",
       "            wordSpacing: 20,\n",
       "            arrowSpacing: 3,\n",
       "\n",
       "        });\n",
       "        const parse = {\n",
       "            arcs: [{\"dir\": \"left\", \"start\": 0, \"label\": \"det\", \"end\": 2}, {\"dir\": \"left\", \"start\": 1, \"label\": \"nn\", \"end\": 2}, {\"dir\": \"left\", \"start\": 2, \"label\": \"nsubjpass\", \"end\": 4}, {\"dir\": \"left\", \"start\": 3, \"label\": \"auxpass\", \"end\": 4}, {\"dir\": \"left\", \"start\": 6, \"label\": \"det\", \"end\": 10}, {\"dir\": \"left\", \"start\": 7, \"label\": \"number\", \"end\": 8}, {\"dir\": \"left\", \"start\": 8, \"label\": \"amod\", \"end\": 10}, {\"dir\": \"left\", \"start\": 9, \"label\": \"nn\", \"end\": 10}, {\"dir\": \"right\", \"start\": 4, \"label\": \"prep_on\", \"end\": 10}, {\"dir\": \"left\", \"start\": 12, \"label\": \"auxpass\", \"end\": 13}, {\"dir\": \"right\", \"start\": 4, \"label\": \"dep\", \"end\": 13}, {\"dir\": \"left\", \"start\": 15, \"label\": \"nn\", \"end\": 16}, {\"dir\": \"right\", \"start\": 13, \"label\": \"prep_with\", \"end\": 16}, {\"dir\": \"right\", \"start\": 4, \"label\": \"dep\", \"end\": 18}, {\"dir\": \"right\", \"start\": 13, \"label\": \"conj_and\", \"end\": 18}],\n",
       "            words: [{\"text\": \"The\", \"tag\": \"DT\"}, {\"text\": \"PCR\", \"tag\": \"NN\"}, {\"text\": \"products\", \"tag\": \"NNS\"}, {\"text\": \"were\", \"tag\": \"VBD\"}, {\"text\": \"electrophoresed\", \"tag\": \"VBN\"}, {\"text\": \"on\", \"tag\": \"IN\"}, {\"text\": \"a\", \"tag\": \"DT\"}, {\"text\": \"3\", \"tag\": \"CD\"}, {\"text\": \"%\", \"tag\": \"NN\"}, {\"text\": \"agarose\", \"tag\": \"NN\"}, {\"text\": \"gel\", \"tag\": \"NN\"}, {\"text\": \",\", \"tag\": \",\"}, {\"text\": \"were\", \"tag\": \"VBD\"}, {\"text\": \"stained\", \"tag\": \"VBN\"}, {\"text\": \"with\", \"tag\": \"IN\"}, {\"text\": \"ethidium\", \"tag\": \"NN\"}, {\"text\": \"bromide\", \"tag\": \"NN\"}, {\"text\": \"and\", \"tag\": \"CC\"}, {\"text\": \"photographed\", \"tag\": \"VBN\"}, {\"text\": \".\", \"tag\": \".\"}]\n",
       "        };\n",
       "\n",
       "        displacy.render(parse, {\n",
       "            uniqueId: 'render_displacy4'\n",
       "            //color: '#ff0000'\n",
       "        });\n",
       "        return {};\n",
       "    });\n",
       "    });\n",
       "    </script>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "example = event_corpus[398][0]\n",
    "bio.render_dependencies(example.sent)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 242,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "defaultdict(list,\n",
       "            {2: [(0, 'det'), (1, 'nn')],\n",
       "             4: [(2, 'nsubjpass'),\n",
       "              (3, 'auxpass'),\n",
       "              (10, 'prep_on'),\n",
       "              (13, 'dep'),\n",
       "              (18, 'dep')],\n",
       "             8: [(7, 'number')],\n",
       "             10: [(6, 'det'), (8, 'amod'), (9, 'nn')],\n",
       "             13: [(12, 'auxpass'), (16, 'prep_with'), (18, 'conj_and')],\n",
       "             14: [],\n",
       "             16: [(15, 'nn')]})"
      ]
     },
     "execution_count": 242,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "example.sent.dependencies\n",
    "example.sent.children"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here the goal is to produce a dictionary that maps the strings \"Child: det->The\" and \"Child: nn->PCR\" to 1.0. \n",
    "\n",
    "To solve this task, implement the feature function below. The passed in `result` is a dictionary you need to populate with more entries, and the `event` argument indicates for which event you need to extract the features. We have already populated the function with some initial code that should get you started.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 243,
   "metadata": {
    "collapsed": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def add_dependency_child_feats(result, event):\n",
    "    \"\"\"\n",
    "    Append to the `result` dictionary features based on the syntactic dependencies of the event trigger word of\n",
    "    `event`. The feature keys should have the form \"Child: [label]->[word]\" where \"[label]\" is the syntactic label\n",
    "    of the syntatic child (e.g. \"det\" in the case above), and \"[word]\" is the word of the syntactic child (e.g. \"The\" \n",
    "    in the case above).\n",
    "    Args:\n",
    "        result: a defaultdict that returns `0.0` by default. \n",
    "        event: the event for which we want to populate the `result` dictionary with dependency features.\n",
    "    Returns:\n",
    "        Nothing, but populates the `result` dictionary. \n",
    "    \"\"\"\n",
    "    index = event.trigger_index # You will need to change this \n",
    "    for child,label in event.sent.children[index]:\n",
    "        word = example.sent.tokens[child]['word']\n",
    "        result[\"Child: \" + label + \"->\" + word ] += 1.0 \n",
    "    return result"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <font color='red'>Assessment 1</font>: Test Feature Function (20 pts)\n",
    "Here we test whether your feature function populates the given dictionary correctly. If the result passes all three tests you get 10 pts. Of course, solutions that just manually populate the result with the specific key value pairs tested below will receive 0 pts as well."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 244,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(True, True, True)"
      ]
     },
     "execution_count": 244,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#! ASSESSMENT 1 - DO NOT CHANGE, MOVE NOR COPY\n",
    "result = defaultdict(float)\n",
    "add_dependency_child_feats(result, example)\n",
    "\n",
    "check_1 = len(result) == 2\n",
    "check_2 = result['Child: det->The'] == 1.0\n",
    "check_3 = result['Child: nn->PCR'] == 1.0\n",
    "(check_1, check_2, check_3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <font color='orange'>Mark</font>:  Your solution to Task 1 is marked with ** __ points**. \n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <font color='blue'>Task 2</font>: Implement Model\n",
    "\n",
    "You are to implement the `predict_event_labels` function below. This function gets as input a list of event candidate objects, and then returns a sequence of corresponding labels. You can implement this function in any way you like, again utilising any library on the docker image. We have populated the cell and function with a simple implementation that uses the scikit-learn logistic regression model. You can use this as a starting point and focus on implementing better feature functions. You can also start from scratch if you like. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'(', ')', ',', '.', 'a', 'activ', 'and', 'cell', 'in', 'of', 'the', 'to'}"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#count the words in training dataset\n",
    "import numpy as np\n",
    "def cal_term_count(dataset):\n",
    "    term_count = {}\n",
    "    for event,_ in dataset:\n",
    "        for token in event.sent.tokens:\n",
    "            word = token['stem']\n",
    "            if word not in term_count.keys():\n",
    "                term_count[word] = 1\n",
    "            else:\n",
    "                term_count[word] += 1\n",
    "    return term_count\n",
    "\n",
    "count_dict = cal_term_count(event_train)\n",
    "x = sorted(count_dict.items(),key=lambda d:d[1], reverse = True)\n",
    "meaningless = set(np.asarray(x)[:12,0])\n",
    "meaningless"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "corpus = []\n",
    "\n",
    "for event, label in event_train:\n",
    "    sent = \"\"\n",
    "    for span in event.argument_candidate_spans:\n",
    "        for i in range(span[0],span[1]):\n",
    "            if event.sent.tokens[i]['stem'] not in meaningless:\n",
    "                sent = sent + event.sent.tokens[i]['stem'] + \" \"\n",
    "    corpus.append(sent)\n",
    "    \n",
    "from sklearn.feature_extraction.text import CountVectorizer,TfidfTransformer\n",
    "vectorizer = CountVectorizer()  \n",
    "transformer = TfidfTransformer()\n",
    "tfidf = transformer.fit_transform(vectorizer.fit_transform(corpus))\n",
    "\n",
    "#word = vectorizer.get_feature_names() \n",
    "#weight = tfidf.toarray()\n",
    "#text = []\n",
    "\n",
    "#for i in range(len(weight)):\n",
    "#    for j in range(len(word)):\n",
    "#        if 0.95 > weight[i][j] > 0.9:\n",
    "#            text.append([word[j],weight[i][j]])\n",
    "#print(tfidf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[defaultdict(float,\n",
       "             {'Child: prep_in->region': 1.0,\n",
       "              'Child: prep_of->expression': 1.0,\n",
       "              'GrandChild: det->the': 1.0,\n",
       "              'GrandChild: nn->factor': 1.0,\n",
       "              'GrandChild: nn->gene': 1.0,\n",
       "              'GrandChild: nn->promoter': 1.0,\n",
       "              'GrandChild: prep_due_to->hypermethylation': 1.0,\n",
       "              'GrandChild: prep_in->cells': 1.0,\n",
       "              'Number of Proteins=5': 1.0,\n",
       "              'Parents: nsubj->is': 1.0,\n",
       "              'Protein lies near Down-regul': 1.0,\n",
       "              'minimum distance between protein and trigger = 2': 1.0,\n",
       "              'trigger is not IN': 1.0,\n",
       "              \"trigger's child is not a Protein\": 2.0,\n",
       "              \"trigger's parent is not a Protein\": 1.0,\n",
       "              'trigger_word=Down-regul': 1.0,\n",
       "              'trigger_word=Down-regulation': 1.0,\n",
       "              'trigger_word=NN': 1.0}),\n",
       " defaultdict(float,\n",
       "             {'Number of Proteins=5': 1.0,\n",
       "              'Protein lies near of': 1.0,\n",
       "              'minimum distance between protein and trigger = 1': 1.0,\n",
       "              'trigger is IN': 1.0,\n",
       "              'trigger_word=IN': 1.0,\n",
       "              'trigger_word=of': 2.0}),\n",
       " defaultdict(float,\n",
       "             {'Child: nn->factor': 1.0,\n",
       "              'Child: nn->gene': 1.0,\n",
       "              'Child: prep_due_to->hypermethylation': 1.0,\n",
       "              'Child: prep_in->cells': 1.0,\n",
       "              'GrandChild: amod->leukemic': 1.0,\n",
       "              'GrandChild: amod->regulatory': 1.0,\n",
       "              'GrandChild: nn->interferon': 1.0,\n",
       "              'GrandChild: num->4': 1.0,\n",
       "              'GrandChild: prep_of->motifs': 1.0,\n",
       "              'GrandParent: nsubj->is': 1.0,\n",
       "              'Number of Proteins=5': 1.0,\n",
       "              'Parents: prep_of->Down-regulation': 1.0,\n",
       "              'Protein lies near express': 1.0,\n",
       "              \"Protein's parent is trigger:nn\": 1.0,\n",
       "              'minimum distance between protein and trigger = 16': 1.0,\n",
       "              'trigger is not IN': 1.0,\n",
       "              \"trigger's child is a Protein\": 1.0,\n",
       "              \"trigger's child is not a Protein\": 3.0,\n",
       "              \"trigger's parent is not a Protein\": 1.0,\n",
       "              'trigger_word=NN': 1.0,\n",
       "              'trigger_word=express': 1.0,\n",
       "              'trigger_word=expression': 1.0}),\n",
       " defaultdict(float,\n",
       "             {'Number of Proteins=5': 1.0,\n",
       "              'minimum distance between protein and trigger = 12': 1.0,\n",
       "              'trigger is IN': 1.0,\n",
       "              'trigger_word=IN': 1.0,\n",
       "              'trigger_word=due': 2.0}),\n",
       " defaultdict(float,\n",
       "             {'Number of Proteins=5': 1.0,\n",
       "              'minimum distance between protein and trigger = 11': 1.0,\n",
       "              'trigger is not IN': 1.0,\n",
       "              'trigger_word=TO': 1.0,\n",
       "              'trigger_word=to': 2.0}),\n",
       " defaultdict(float,\n",
       "             {'Number of Proteins=5': 1.0,\n",
       "              'minimum distance between protein and trigger = 9': 1.0,\n",
       "              'trigger is IN': 1.0,\n",
       "              'trigger_word=IN': 1.0,\n",
       "              'trigger_word=of': 2.0}),\n",
       " defaultdict(float,\n",
       "             {'GrandParent: prep_in->Down-regulation': 1.0,\n",
       "              'Number of Proteins=5': 1.0,\n",
       "              'Parents: nn->region': 1.0,\n",
       "              'minimum distance between protein and trigger = 4': 1.0,\n",
       "              'trigger is not IN': 1.0,\n",
       "              \"trigger's parent is not a Protein\": 1.0,\n",
       "              'trigger_word=NN': 1.0,\n",
       "              'trigger_word=promot': 1.0,\n",
       "              'trigger_word=promoter': 1.0}),\n",
       " defaultdict(float,\n",
       "             {'Child: det->the': 1.0,\n",
       "              'Child: nn->-abl': 1.0,\n",
       "              'Child: nn->bcr': 1.0,\n",
       "              'GrandParent: advcl->is': 1.0,\n",
       "              'GrandParent: xcomp->shown': 1.0,\n",
       "              'Number of Proteins=5': 1.0,\n",
       "              'Parents: nsubjpass->shown': 1.0,\n",
       "              'Parents: xsubj->aberration': 1.0,\n",
       "              'Protein lies near transloc': 2.0,\n",
       "              \"Protein's parent is trigger:nn\": 2.0,\n",
       "              'minimum distance between protein and trigger = 34': 1.0,\n",
       "              'trigger is not IN': 1.0,\n",
       "              \"trigger's child is a Protein\": 2.0,\n",
       "              \"trigger's child is not a Protein\": 1.0,\n",
       "              \"trigger's parent is not a Protein\": 2.0,\n",
       "              'trigger_word=NN': 1.0,\n",
       "              'trigger_word=transloc': 1.0,\n",
       "              'trigger_word=translocation': 1.0}),\n",
       " defaultdict(float,\n",
       "             {'GrandParent: xcomp->shown': 1.0,\n",
       "              'Number of Proteins=5': 1.0,\n",
       "              'Parents: aux->aberration': 1.0,\n",
       "              'minimum distance between protein and trigger = 30': 1.0,\n",
       "              'trigger is not IN': 1.0,\n",
       "              \"trigger's parent is not a Protein\": 1.0,\n",
       "              'trigger_word=TO': 1.0,\n",
       "              'trigger_word=to': 2.0}),\n",
       " defaultdict(float,\n",
       "             {'Child: amod->causative': 1.0,\n",
       "              'Child: amod->genetic': 1.0,\n",
       "              'Child: aux->to': 1.0,\n",
       "              'Child: cop->be': 1.0,\n",
       "              'Child: det->the': 1.0,\n",
       "              'Child: prep_in->leukemia': 1.0,\n",
       "              'Child: xsubj->translocation': 1.0,\n",
       "              'GrandChild: abbrev->CML': 1.0,\n",
       "              'GrandChild: amod->chronic': 1.0,\n",
       "              'GrandChild: amod->myeloid': 1.0,\n",
       "              'GrandChild: det->the': 1.0,\n",
       "              'GrandChild: nn->-abl': 1.0,\n",
       "              'GrandChild: nn->bcr': 1.0,\n",
       "              'GrandParent: advcl->is': 1.0,\n",
       "              'Number of Proteins=5': 1.0,\n",
       "              'Parents: xcomp->shown': 1.0,\n",
       "              'minimum distance between protein and trigger = 25': 1.0,\n",
       "              'trigger is not IN': 1.0,\n",
       "              \"trigger's child is not a Protein\": 7.0,\n",
       "              \"trigger's parent is not a Protein\": 1.0,\n",
       "              'trigger_word=NN': 1.0,\n",
       "              'trigger_word=aberr': 1.0,\n",
       "              'trigger_word=aberration': 1.0})]"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from sklearn.linear_model import LogisticRegression\n",
    "import numpy as np\n",
    "from collections import defaultdict\n",
    "from sklearn.feature_extraction import DictVectorizer\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "\n",
    "# converts labels into integers, and vice versa, needed by scikit-learn.\n",
    "label_encoder = LabelEncoder()\n",
    "\n",
    "# encodes feature dictionaries as numpy vectors, needed by scikit-learn.\n",
    "vectorizer = DictVectorizer()\n",
    "\n",
    "def event_feat(event):\n",
    "    \"\"\"\n",
    "    This feature function returns a dictionary representation of the event candidate. You can improve the model \n",
    "    by improving this feature function.\n",
    "    Args:\n",
    "        event: the `EventCandidate` object to produce a feature dictionary for.\n",
    "    Returns:\n",
    "        a dictionary with feature keys/indices mapped to feature counts.\n",
    "    \"\"\"\n",
    "    result = defaultdict(float)\n",
    "    trigger = event.trigger_index\n",
    "    result['trigger_word=' + event.sent.tokens[trigger]['word']] += 1.0\n",
    "    result['trigger_word=' + event.sent.tokens[trigger]['stem']] += 1.0\n",
    "    result['trigger_word=' + event.sent.tokens[trigger]['pos']] += 1.0\n",
    "\n",
    "    for child,label in event.sent.children[trigger]:\n",
    "        word = event.sent.tokens[child]['word']\n",
    "        result[\"Child: \" + label + \"->\" + word ] += 1.0 \n",
    "        if event.sent.is_protein[child]:\n",
    "            result[\"trigger's child is a Protein\"] += 1.0\n",
    "        else:\n",
    "            result[\"trigger's child is not a Protein\"] += 1.0\n",
    "        #result[\"Trigger_children:\" + str(child - trigger)] += 1.0\n",
    "        for grandchild,grandlabel in event.sent.children[child]:\n",
    "            grandword = event.sent.tokens[grandchild]['word']\n",
    "            result[\"GrandChild: \" + grandlabel + \"->\" + grandword ] += 1.0 \n",
    "            \n",
    "    for parent,label in event.sent.parents[trigger]:\n",
    "        word = event.sent.tokens[parent]['word']\n",
    "        result[\"Parents: \" + label + \"->\" + word ] += 1.0 \n",
    "        if event.sent.is_protein[parent]:\n",
    "            result[\"trigger's parent is a Protein\"] += 1.0\n",
    "        else:\n",
    "            result[\"trigger's parent is not a Protein\"] += 1.0\n",
    "\n",
    "        #result[\"Trigger_parents:\" + str(parent - trigger)] += 1.0\n",
    "        for grandparent,grandlabel in event.sent.parents[parent]:\n",
    "            grandword = event.sent.tokens[grandparent]['word']\n",
    "            result[\"GrandParent: \" + grandlabel + \"->\" + grandword ] += 1.0 \n",
    "    \n",
    "\n",
    "    if event.sent.tokens[trigger][\"pos\"]!=\"IN\" :\n",
    "        result[\"trigger is not IN\"] += 1.0\n",
    "    elif event.sent.tokens[trigger][\"pos\"]==\"IN\" :\n",
    "        result[\"trigger is IN\"] += 1.0\n",
    "\n",
    "\n",
    "    z = event.sent.mentions\n",
    "    result[\"Number of Proteins=\" + str(len(z))] += 1.0\n",
    "    min_distance = float(\"inf\")\n",
    "    for i in range(len(z)):\n",
    "        begin = z[i][\"begin\"]\n",
    "        end = z[i][\"end\"]\n",
    "        label = z[i][\"label\"]\n",
    "        word = event.sent.tokens[trigger]['stem']\n",
    "        if (end == trigger) or (end == trigger - 1) or (begin == trigger + 1) or (begin == trigger + 2):\n",
    "            result[\"Protein lies near \"+ word] += 1.0\n",
    "\n",
    "        if begin > trigger:\n",
    "            if begin - trigger  <= min_distance:\n",
    "                min_distance = begin - trigger \n",
    "        \n",
    "        \n",
    "        for j in range(begin,end):\n",
    "            for parent,label in event.sent.parents[j]:\n",
    "                if parent == trigger:\n",
    "                    result[\"Protein's parent is trigger:\" + label] += 1.0\n",
    "                        #result[\"distance\"+str(parent-begin)]+=1.0\n",
    "\n",
    "            for child,label in event.sent.children[j]:\n",
    "                if child == trigger:\n",
    "                    result[\"Protein's child is trigger:\" + label] += 1.0\n",
    "                        #result[\"distance\"+str(child-begin)]+=1.0\n",
    "\n",
    "    result[\"minimum distance between protein and trigger = \" + str(min_distance)] += 1.0\n",
    "    return result\n",
    "# We convert the event candidates and their labels into vectors and integers, respectively.\n",
    "train_event_x = vectorizer.fit_transform([event_feat(x) for x,_ in event_train])\n",
    "train_event_y = label_encoder.fit_transform([y for _,y in event_train])\n",
    "\n",
    "# Create and train the model. Feel free to experiment with other parameters and learners.\n",
    "lr = LogisticRegression(C=1,class_weight='balanced')\n",
    "lr.fit(train_event_x, train_event_y)\n",
    "\n",
    "def predict_event_labels(event_candidates):\n",
    "    \"\"\"\n",
    "    This function receives a list of `bio.EventCandidate` objects and predicts their labels. \n",
    "    It is currently implemented using scikit-learn, but you are free to replace it with any other\n",
    "    implementation as long as you fulfil its contract.\n",
    "    Args:\n",
    "        event_candidates: A list of `EventCandidate` objects to label.\n",
    "    Returns:\n",
    "        a list of event labels, where the i-th label belongs to the i-th event candidate in the input.\n",
    "    \"\"\"\n",
    "    event_x = vectorizer.transform([event_feat(e) for e in event_candidates])\n",
    "    event_y = label_encoder.inverse_transform(lr.predict(event_x))\n",
    "    return event_y\n",
    "\n",
    "[event_feat(x) for x,_ in event_train][:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It is useful to inspect the performance of your model, and see where it makes errors, both on the training set (to check for underfitting) and the development set. We have provided you with utility functions to help with this inspection. Note that you don't have to use these utilities, or the cells below, but it can help you to improve your model, and also with the error analysis and description of the approach in Task 3. \n",
    "\n",
    "First, we give you a breakdown of precision, recall and F1 on different event types:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Label</th>\n",
       "      <th>Gold</th>\n",
       "      <th>Guess</th>\n",
       "      <th>Precision</th>\n",
       "      <th>Recall</th>\n",
       "      <th>F1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Binding</td>\n",
       "      <td>180</td>\n",
       "      <td>211</td>\n",
       "      <td>0.601896</td>\n",
       "      <td>0.705556</td>\n",
       "      <td>0.649616</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Gene_expression</td>\n",
       "      <td>377</td>\n",
       "      <td>423</td>\n",
       "      <td>0.756501</td>\n",
       "      <td>0.848806</td>\n",
       "      <td>0.800000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Localization</td>\n",
       "      <td>71</td>\n",
       "      <td>64</td>\n",
       "      <td>0.781250</td>\n",
       "      <td>0.704225</td>\n",
       "      <td>0.740741</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Negative_regulation</td>\n",
       "      <td>210</td>\n",
       "      <td>249</td>\n",
       "      <td>0.550201</td>\n",
       "      <td>0.652381</td>\n",
       "      <td>0.596950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Phosphorylation</td>\n",
       "      <td>32</td>\n",
       "      <td>37</td>\n",
       "      <td>0.837838</td>\n",
       "      <td>0.968750</td>\n",
       "      <td>0.898551</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Positive_regulation</td>\n",
       "      <td>570</td>\n",
       "      <td>694</td>\n",
       "      <td>0.582133</td>\n",
       "      <td>0.708772</td>\n",
       "      <td>0.639241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Protein_catabolism</td>\n",
       "      <td>21</td>\n",
       "      <td>23</td>\n",
       "      <td>0.739130</td>\n",
       "      <td>0.809524</td>\n",
       "      <td>0.772727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Regulation</td>\n",
       "      <td>188</td>\n",
       "      <td>208</td>\n",
       "      <td>0.543269</td>\n",
       "      <td>0.601064</td>\n",
       "      <td>0.570707</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Transcription</td>\n",
       "      <td>112</td>\n",
       "      <td>125</td>\n",
       "      <td>0.544000</td>\n",
       "      <td>0.607143</td>\n",
       "      <td>0.573840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>[All]</td>\n",
       "      <td>1761</td>\n",
       "      <td>2034</td>\n",
       "      <td>0.622911</td>\n",
       "      <td>0.719478</td>\n",
       "      <td>0.667721</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Label  Gold  Guess  Precision    Recall        F1\n",
       "0              Binding   180    211   0.601896  0.705556  0.649616\n",
       "1      Gene_expression   377    423   0.756501  0.848806  0.800000\n",
       "2         Localization    71     64   0.781250  0.704225  0.740741\n",
       "3  Negative_regulation   210    249   0.550201  0.652381  0.596950\n",
       "4      Phosphorylation    32     37   0.837838  0.968750  0.898551\n",
       "5  Positive_regulation   570    694   0.582133  0.708772  0.639241\n",
       "6   Protein_catabolism    21     23   0.739130  0.809524  0.772727\n",
       "7           Regulation   188    208   0.543269  0.601064  0.570707\n",
       "8        Transcription   112    125   0.544000  0.607143  0.573840\n",
       "9                [All]  1761   2034   0.622911  0.719478  0.667721"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This line calls your function to produce labels for the test set\n",
    "event_dev_guess = predict_event_labels([x for x,_ in event_dev[:]])\n",
    "# This line produces a confusion matrix\n",
    "cm_dev = bio.create_confusion_matrix(event_dev,event_dev_guess)  \n",
    "# This line turns the confusion matrix into a evaluation table with Precision, Recall and F1 for all labels.\n",
    "bio.full_evaluation_table(cm_dev)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It is useful to inspect [bio.py](/edit/statnlpbook/bio.py) to see how we define precision, recall and F1 score in this context.\n",
    "\n",
    "You can also display a confusion matrix to identify what types of errors you are currently making. Notice that the matrix ignores the \"None\"-\"None\" cell as its counts would overpower all other counts (try removing the `outside_label` argument). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAUgAAAEYCAYAAAA+mm/EAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzsnXe8XFXZtq87BQgloSMdpLdQEpCqSFeRYuNFQAQUBKTI\niwp+FsRXQRGRIkjoVaoUUXozdEghoQpSlCK9SyAJ9/fHWpuzz2TKnpkzZ+bMWRe//Tuzy1rrmQnz\nzCrPuh/ZJpFIJBKzMqTdBiQSiUSnkhxkIpFIVCA5yEQikahAcpCJRCJRgeQgE4lEogLJQSYSiUQF\nkoNMJBKJCiQHmUgkBjSShkqaJOmaeD6/pBslPRH/zpd79nBJT0p6XNLWtepODjKRSAx0DgIezZ0f\nBtxsewXg5niOpFWB/wFWA7YBTpY0tFrFw1pibqItjJxvfi+82JINlR01x/A+tqb/+HDmRw2XnW1o\n6iM0wsSJE161vVCz9QwdubQ94/2y9/z+K9fb3qZaeUlLAF8AfgkcEi9vD2waX58D3Ab8MF6/yPYH\nwNOSngTWA+6uVH9ykF3Ewostye8uur6hslut8ok+tqb/eOGN8l+wIiw234g+tGTwMGK4nu2Lejxj\nGrOv/D9l702bdOLKkh7IXRpne1zJY78HfgDMk7u2iO0X4+v/AIvE14sD9+Seey5eq0hykIlEon0I\nGFJxlPuq7bEVi0rbAi/bniBp03LP2LakhgUnkoNMJBJtRNUcZC02AraT9HlgDmCkpPOBlyQtavtF\nSYsCL8fnnwfyc1BLxGsVSRMwVZA0U9JkSQ9Kmihpw3h9MUmX1VnXNyWdFF9/R9I3WmFzIjHgkMof\nNbB9uO0lbC9DWHy5xfauwNXA7vGx3YGr4uurgf+RNLukZYEVgPuqtZF6kNV53/ZaADEk4CjgM7Zf\nAL7SaKW2/9hH9iUSAxs11YOsxNHAJZL2Ap4FvgZg+2FJlwCPADOA/W3PrFZRcpDFGQm8ASBpGeAa\n26tL+iawHTAnsBxwhe0fxOf2AA4H3gQeBD6I148A3rX9W0m3AfcCnwXmBfayPV7SnMDZwOrA48Bi\nhH/Q/KR1IjHw6QMHafs2wmo1tl8DNq/w3C8JK96FSA6yOiMkTSbMbywKbFbhubWAtQkO8HFJJxJ+\noX4OjAHeAm4FJlUoP8z2enEu5WfAFsB+wBu2V5W0OjC5j95TItFBCNS5M33JQVYnP8TeADg3OqtS\nbrb9VnzuEWBpYEHgNtuvxOsXAytWaOfP8e8EYJn4emPgeADbD0maUq6gpL2BvQEWWrRqxEIi0XlU\nX8VuO53rujsM23cTnF654NgPcq9nUv8PT1a+7rK2x9kea3vsyPkWqLPZRKLdCIYOLX90AMlBFkTS\nysBQ4LWCRe4FPiNpAUnDga/W2eSdxMnluEVqjTrLJxKdjwhD7HJHB5CG2NXJ5iAh/FPubnumioUg\nvBgXY+4mLNLUO4d4MnBOHLI/BjxMmMtMJLqIlqxi9xnJQVbBdtl/OdvPEFaXsX02YbU5u7dt7vVZ\nwFllyh+Re71p7vWr9MxBTgN2tT1N0nLATYSQhUSiu0gOMtEAcwK3xuG5gP1sf9hmmxKJvqVgUHi7\nSA6yQ7H9DlBxH2oi0TWkHmSiPxg1x/CGVXmW3f/yptp++g9fbris3bCWAADD2ihZNvOjxm0fOqR9\nPaePmrC7b0lxkIlEIlGeFAeZSCQSlYir2OWOWiWlOSTdF8VkHpb083j9CEnPR6GZyXGHWlamrpQL\nqQeZSCTaS+ND7A+AzWy/Gxcz75B0bbx3nO3f9mqmd8qFxYCbJK1YTbAi9SATiUT7UOM9SAfejafD\n41FtcvXjlAu2nwaylAsVSQ4ykUi0FQ0ZUvYoVDZkNJxMEMW90fa98dYBkqZIOjOX1XBx4N+54jVT\nLrTcQUpaRNKFkp6SNEHS3ZJ2bHW7nYykIyVt0W47Eol2I0BS2QNYUNIDuWPv0vK2Z0ZBmSWA9aKY\nzCnAJwkqWy8CxzZqX0vnIBXe5ZXAOba/Hq8tTdBPHHBIGmZ7RrP12P5pX9iTSAx4JFQ53KlqTpo8\ntt+UdCuwTX7uUdJpwDXxtONSLmwGfJhX0Lb9rO0TY9f4GEn3x67wPgCSNpV0m6TLJD0m6YLoaJE0\nRtLtsSd6fcw3URZJy0m6Lj47PopNIOmqLN2BpH0kXRBf3ybp+Ljq9ZCk9eL1IySdJ+lO4Lwqdi8q\n6e+58pvEZ8+O51MlfS8+e7akr8TXmyskPZ8ahwOzx+vPSPq5QqqHqZn9Zd7n3tkv7CuvvtLcv1Yi\n0QaGDBlS9qiFpIUkzRtfjwC2BB4r8Qs7Ag/F1x2XcmE1YGKFe3sBb9leNzqFOyXdEO+tHcu+QFC1\n2UjSvcCJwPa2X5G0E0EZeM8K9Y8DvmP7CUmfIog/bEbQTrxT0tPA/wLr58rMaXstSZ8GziTutwZW\nBTa2/X7s5pez+0vA9bZ/qZCMfE5CF39x26sDZP+YGZLmIOzj3tz2PySdC+xLSGUJ4Rd0HUn7AYcC\n3yp9kzEN5jiAMWPGdkr0byJRGDW+1XBRgqDLUEJn7xLb18QOzVqEBZtngH1gAKRckPQHghDshwTh\nhdFZTwoYRfDoHwL32X4ulplMEHB4k+Cwbowf6FDC/EK5duYGNgQuzX34swPYfknSTwkK3zvafj1X\n9E/xmb9LGplzaFfbzpIvb1XB7vuBMxXCDa60PVnSU8AnFRTG/wrcQG9WAp62/Y94fg6wPz0OMi+k\n+6Vy7zWRGMio+hC7KranEDpTpdd3q1Kmo1IuPAx8vAfN9v6SFgQeAP4FHGC7V6Z7hfy25QRoBTxs\ne4MC7Q4B3szUwMuwBkHXcbGS66U9sOz8vbyJ5eyOtn8a+AJwtqTf2T5X0prA1sB3CPqOlXq85WhY\nSDeRGCgUGU63i1Zbdgswh6R9c9fmjH+vB/aNPS4krShprip1PQ4spJD6AEnDJa1W7kHbbwNPS/pq\nfFbRURHnFj9H+OU5NM5FZOwUn9mYMIwup79Y1m6FxaeXbJ8GnA6sE38Mhti+HPgxsE6Z97SMpOXj\n+W7A7VU+g0Si61DlVey209JeiW1L2gE4TtIPgFcIvbEfApcShs4TFT6NV4AdqtT1YRzWniBpVLT9\n94Reajl2AU6R9GNCAOlFkh4DTgP2sP2CpP8lDIuzZFzTJE2Kz1fq6Z1ewe5Nge9Lmg68C3yDEGN1\nlvTxVoHDS97TNIXMh5dKGkYYpqeUsInBg2h4iN0fqFkllW5BIf3qoQM5reqYMWN9572NmT+Q1Xxe\neadxmcyFR87eVNuDVc1nrtmHTCgaglON4Qsu5/m2O6rsvVfO2qlP2miGNK+VSCTaS+d2IAe+g4wr\n4xuVXD4+pjsoTD71wWDknyc2t0j+4pvTGi77iVHN9eLmnXN4U+WboYNHh1UZ0imGq7MXaQa8g7S9\nf7ttSCQSjSGUHGQikUhUpEM6s+VIDjKRSLSPDh9id65ldSLp3dpPNVz3MzGmEUl3NVjHj0rOG6on\nkeg2NERlj06gaxxkf2F7wwaL9nKQTdSTSHQVnRwo3tUOUtIykm5RUN25WdJS8foikq5QyGXxoKQN\n4/UrFdR/HlYZ7bn4zLvx75HqyXnxvKSzKtUh6WhgRHz2gpJ6pKAOlCn+ZLt5KqoaJRLdgqRm1Hwq\n5aSZX9KNkp6If+fLlakrJ01XO0iC+s85tkcDFwAnxOsnALfbXpOw/S/bjbOn7TGEfNQHSlqgUsW2\nfxr3em8KvA6cVKkO24cB79tey/YuJVV9iaD6syawBXCMeuSa1gYOJqgJfZJZw5kSiQFPEz3ILCfN\nmoTv0DaS1gcOA262vQJwczwvzUmzDXCyghJQRbrdQW4AXBhfn0dQEoIge3YKfKxInO25PlDSg8A9\nBGHNFapVHnt05wO/sz2hkTqiTX+KdrxE2Iu9brx3n+3nbH8EZKpGpTYkPcjEgKbROcgqOWm2Jyhj\nEf9mW5hTTppGUVAR2gLYIP4iTQLmqFHsCOC5LCi9wTqqUU7VqBe2x9kea3vsQgsu1ERTiUT/I4Wg\n9XIHBVIuqHxOmkVsZ1KI/wEWia87LydNm7mL0KWGIF4xPr6+mSBMm33Aowi6jm/Y/q+Cevf6pZXl\nkfRFgjM8MHe5Wh3TFRWAShgP7BTtWAj4NDVUjhOJ7qH88DoOsV/NfvzjMa60dIWcNPn7pnqmw6p0\nk4OcU9JzueMQ4ABgD0lTCFJiB8VnDwI+K2kqQYx2VeA6YJikR4GjCUPkahxC+PW5Ly6+HFmjjnHA\nlGyRJscVwBTgQYI83A9s/6eRDyCRGIhU6UEWxvabBBHsbYCXsnn8+Pfl+FjdOWmSmk8X0YyaT7Pq\nLi+9/UHthyrQ7F7s6TMbt322Yc31EZr5/gzkoIQRw9UnSjsjFl3Ry+5xUtl7jx61ddU24ohrekzY\nNYKg2P9r4DPAa7aPlnQYML/tHyjox15ImHdcjDCSXKFa2oW0kyaRSLSVJn4nKuWkuRu4RNJehNQu\nX4MBkJMmkUgkeqHGlYWq5KR5Ddi8QpmOykmTGCA0K3+16LyNL9b/5aEXmmp79Cfmrf1QBZZecM7a\nD1WhmZmJoU2OsJuZFukUubOk5pNIJBJV6OSp2OQgE4lE+2hiiN0fJAeZSCTahkgOMpFIJCrSyeFO\nyUEmEon20eFD7JrLR5Is6djc+aGSjuhrQzRABWUVJNUeKvDM13PnYyWdUK1MIjEYCKvYze+kaRVF\n1tc/AL6kqKjdQvpFUFZSO3rNywAfO0jbD9g+sPLjicTgQSp/dAJFHOQMwj7i75XekLSQpMsl3R+P\njXLXb4wilqdLelY9KQvqFZS9SNIXcm2eLekrUdzhmNjuFEn7VHoDCuKz4yVdTYiiR9KuUWxzsqRT\nM104SXtJ+ke8d5qkk/Lt5uqcJcVD7CmOlzQxHpmTPxrYJLb1vWjPNbHM/PEzmSLpHkmj4/UjJJ2p\nIJr7lKSyDlVJ7iwxkKmu5tN2ikZo/gHYJare5DkeOM72usCXgdPj9Z8Bt9heDbgMWCpXpl5B2YuJ\nW4UkzUaIkP8rsBfwVmx7XeDbkpat8h7WAQ6yvaKkVYCdgI2iEsjM+P4WA35CUOHZCFi50KfTw8vA\nlrbXifVnw+jDgPHx/R1XUubnwKQo6vsj4NzcvZWBrQl7R39WTg0oyZ0lBjJhFbsxRfH+oNBw0/bb\nks4lSHu9n7u1BbBqbhVqpKS5CSKwO8ay10l6I1fmQEk7xteZoOxrVZq/Fjhe0uwEpY6/235f0lbA\n6FyvblSs6+kK9dwXRTIhONkxwP3R9hEE57YeQWn8dQBJlwIrVrGtlOHASZIyp1uk7MaEHxds3yJp\nAUkj472/2v4A+EDSywRdu+fqsCeR6HgaHU5LWpLQoViEIGk2zvbxcY3k20A2pPqR7b/FMocTOlcz\ngQNtX1+tjXrm434PTATOyl0bAqxve1qJ4ZXe0Kb0CMr+V9Jt1BCUtT0tPrc1oVd2UVYdcECtN5jj\nvbwphFQMh5fYtwOVmUHscUsaAsxW5pnvAS8R0icMAaaVeaYeagrmJhIDmuZWsWcA/2t7oqR5gAmS\nboz3jrP9215N9U65sBhwk6QVqwlWFO7Hxl7VJQTvm3EDQXMxM2Ct+PJOeobFWwFZ0pxGBGUhDLP3\nADYhaC4CXA/sm5WRtKKkuQq+nZuBr0haOJadX9LSwP3AZyTNFxdzvpwr8wyh1wmwHaG3WMoo4MWY\nImE3IMt38Q4wTwVbxhPEfLMfkFdtv13wfSQSA5pmVrFtv2h7Ynz9DvAo1RXCW55y4Vggv5p9IDA2\nLjA8AnwnXv85sFUMf/kqQfb8HRoTlIXgiD8D3GT7w3jtdMKCy8TYzqkUnzJ4BPgxcIOCmO6NwKK2\nnwd+RVD0vpPgFLN8NacRnOeDhFw375XWC5wM7B6fWTn3zBRgpkL2tdLFriOAMdGOo4Hdi7yHRKJb\nGCKVPSiQciFD0jIEZZ9746UDol86Uz1ZDetOudASwdw4XzjT9gxJGwCnxMWQjkfS3LbfjT3IK4Az\nbV/RbruK0IxgbjsZyGo+M5tQ1Bna5EptO9V8+kowd9TSq3j9H55d9t4N+69fqI247nE78Evbf5a0\nCPAqYV7yF4TOz54xIuUe2+fHcmcA19q+rFLdrZrTWoogWDkE+JAwYTpQOELSFoS50RuAK9tsTyLR\n1TTjq+MU2+XABbb/DBCzg2b3TwOuiad1p1xoiYO0/QRlhCxbjaQ1COld83xg+1NF67B9aN9a1X+8\nP30mjz7f2PRlswOJ5T8xd8NlF5trRFNtN9MLbKYHCLDtKXc3XPaafTdoqu0n/jNLKG5hRo7onPW+\nRnuzCqvBZwCP2v5d7vqiuayGOwLZTrergQsl/Y6wSLMCNRLkdc6n1AfYnkpIIJ5IJAYAAoY2vm1m\nI8Ji6FSF1K8QYol3jgvGJqwj7AMp5UIikRho9KR4rRvbdxB8bCl/q1ImpVxIJBIDA9H8YlUr6Yz9\nPF2C+kn5KJHoJrphL3aiGP2lfJRIdAWVlHwGkppPojjVlI+WkXRLDF69WdJS8frZkk6QdFdU7ckr\nBn1fPWpFP++/t5FI9B9DpbJHJ5AcZN9TSfnoRML+79HABfQo/UBIgL4xsC1hN022RXMFwlaotQi7\nbT7dYtsTiX4lm4Msd3QCyUH2MXEfdaZ8lGcD4ML4+jyCQ8y40vZHcQvkIvHaVvGYRBAJWZngMHuR\n14N84/VqokiJRAcSV7HLHZ1AWsVuDeWUj6qRV+1R7u9Rtk+tVtD2OMKwnlVHr933+0YTiRbTKQsy\n5Ug9yBZQQfnoLoLUEgT1nvE1qrke2DPuM0XS4pn6UCLRLaQh9uClVPnoAGCPqNqzG3BQtcK2byAM\nye+WNJWgzF5JMi2RGLCowtEJpCF2H2J77tzrl4A5c+fPApuVKfPNKnUcT0hrkUh0JVIKFE8kEomK\nNBooLmlJSbdKekQhCeBB8fr8CkkDn4h/58uVOVzSk5Iel7R1TduaemeJRCLRBKK8WO6QYqvYWcqF\nVQnZCfaPaRUOA262vQIhe8BhMEvKhW2AkxWzmVYiDbG7iBHDh7LK4iNrP9hhjFl2vtoPtYh3p81o\nqvy1+7ckfXshVl6sC6akm8hJEyXNXoyv34mZChYnpFbYND52DnAb8ENyKReApyVlKRcqatYlB5lI\nJNpGDbmzBSXlJfLHxbC2WevpnXJhkZwe5H/oiS1enN5pXmqmXEgOMpFItJUqHchX60i5cDlwcExR\n/fE925bUcHxwcpCJRKJtNLuKXS7lAvBSpiouaVFCzntoIOVCWqRJJBJtZeiQ8kctKqVcIKRWyLKD\n7g5clbv+P5Jml7QsBVIutMVBSpopabKkhyRdKmnOqHbzUO3SDbe5qaRraj/ZUN23Sao6FJB0sKQ5\nc+d/k9R4Or5EogsQVdO+1iJLubBZ9CeTJX2eIPiypaQngC3iObYfJuxwe4SQgrpjUy68n6WBjXmw\nvwP8uXqR9iJpaK0PswYHA+cD/wWw/fk+MSyRGOAMbXCEXSXlAsDmFcrUlXKhE4bY44Hl4+uhkk6L\nQZ83SBoBIGktSfdEXcQrssBPSQfGINEpki6K146QdJ6ku2OgaD7l7NySLpP0mKQLYhcdSZtLmiRp\nqkKi8dnj9Wck/VrSROCw+Jd4b4X8ee76KVFd5+FMw1HSgYQsardKujVX94Lx9SGxN/2QpIPjtWUk\nPVru80gkugVV6D0W7EG2nLY6SEnDgM8BU+OlFYA/2F4NeBP4crx+LvDDqKU4FfhZvH4YsHa8/p1c\n1aMJ2/o2AH4qabF4fW1CT25V4JPARpLmAM4GdrK9BqFXvW+urtdsrxN/ed5SyJYGsAfl1Xr+X1x5\nGw18RtJo2ycALwCftf3Zks9gTKzrU4Rg129LylLmVvo88uU/ljt75dVXypiTSHQ2jc5B9gftMmOE\nQprGB4B/ESZaAZ62naVvnAAsoyA8O6/t2+P1c4BMOHYKcIGkXQlR9RlX2X7f9qvArYRgUID7bD9n\n+yNgMrAMsFJs9x9l6ge4OPf6dILgxFBgJ3r0HfN8LfYsJxEi9let8VlsDFxh+z3b7xKmGjap9HmU\nFrY9zvZY22MXWnChGk0lEp1Fp6v5tH0OMiOOdvO6iDOBWkPKLxCc2ReB/ydpjXi9NO4pOy+tv8j7\nfy/3+nJC7/UWYILtXgq1cWXsUGBd229IOhuYo0Ablaj380gkBhaqGgfZdjqkI1sZ228Bb0jKelW7\nAbdLGgIsaftWwjaiUUCmhLO9pDkkLUDYcnR/lSYeJ/RUs3nQ3YDbyz1oexpBp/EUyg+vRxIc6luS\nFiFMH2S8Q3m5svHADnElfy5gR2prRSYSXUG2k6ZTc9IMlEDx3YE/xjCZpwhzdkOB8+MQXMAJtt+M\nPdEphKH1gsAvbL8gacVyFdueJmkP4NI4J3o/8McqtlxAcGI3lKnrQUmTgMeAfwN35m6PA66T9EJ+\nHtL2xNjTzOKxTrc9KW6dSiS6nk4ZTpdDdnep9CvkoX7X9m9bVP+hwCjbP2lF/c0wZsxY33nvA7Uf\nTHzMW/+d3lT5UXMO7yNLBhYjhmtCkW2AtVhypTV88Liryt47dNPl+qSNZhgoPciOQNIVwHKUEb5N\nJBIN0OGCuV3nIG0f0cK6d2xV3YnEYCSsYrfbisp0nYNMDD7Oe+DZhsvuNnbpptqeb93vNlz2jftP\naqrtGTM/arjssI7xSmJIx2SgmZVO+ZQSicQgJKj5NB4oHne+vZzXcYi76Z4v2Z+d3asr5ULqQSYS\nibaRBYo3wdnASYTddnmOK12oVe+UC4sBN0lasZrGQupBJhKJttLMXmzbfwdeL9jUxykXbD8NZCkX\nKttWsOJEIpHoc0KgePmDmHIhd+xdR9UHRBGbM9WT1XBxQnxyRs2UCw07SJXRdGygjtNjtxdJPyq5\nd1ejtvUnKqBjGZ/5eu58rKQTWm9dItHhqGoP8tVMZyAeZfPRlOEUghjNWoSkXsc2al4zPcj3ba9l\ne3XgQ3qr6RTC9rdsPxJPf1RyryXp4uJumf5mGeBjB2n7AdsHtsGORKKjaFIwtyy2X7I9M4rSnEbP\nMLptKRc+1nSsoG04l6S/SnowXt8pXr8t9qaOJir8KAjoIund+PciSV/IGpJ0tqSvSBoq6RhJ98eu\n9D6VjFNQEx8v6WqCmjCSdpV0X2zz1KjQg6S9JP0j3jtN0kn5dnN1vlumnWViOxPjkTn5o4FNYlvf\nU07dXCHJ+ZXxPdwjaXS8fkQcHtwm6SkFTclEousYovJHoyjkocnYEchGeHWnXGi6N6UeTcfr1Fvb\nUMC9km4ndHdfsP2FWGZUvg7bh0n6bqnCT+Ri4GvAXyXNRlAK3hfYC3jL9roKArd3SrohTr6WYx1g\nddtPS1qFIFe2ke3pkk4GdpF0E/CT+Ow7BNWeB+v4OF4Gtoz7u1cA/gSMJehWHmp72/j+N82V+Tkw\nyfYOkjYjrMZln8PKwGcJIhePSzrFdq+9cXFeZm+AJZdaqg5TE4n2I5oTppD0J4IgzYKSniOobW2q\noNtq4BlgHwgpFyRlKRdm0OKUC5mmI4Qe5BkEx3WF7fei8Zm24XXAsZJ+DVxjux61mmuB46MT3Ab4\nu+33JW0FjM716kYRfhEqOcj7cs5zc2AMcL/CP84IgnNbD7jd9uvR/kuBsiIXFRgOnBT/cWYWLLsx\nUQjX9i2SFpA0Mt77a0xy/oGklwn5fZ/LF47zMuMg7MWuw9ZEoiNQc8PpnctcPqPMtez5ulIuNOMg\nK2k6ljPqH5LWAT4P/J+km20fWaSR2Bu7Ddia0Ou7KGsOOMD29QXtzes6CjjH9uEl9u9QpfwM4pSE\ngtTabGWe+R7wErBmfHZaQdsq0Yh+ZSIxcIiLNJ1KX4f5lNU2VEh58F/b5wPHEIawpUxXyHFbjosJ\nQ/esNwpBl3HfrIykFWObRbgZ+IqkhWPZ+SUtTZA6+4yk+eLUQT7FwTOEXifAdoTeYimjgBfj5PBu\nBEk2qKwFCeEz2yXasSlh5e7tgu8jkRjQDCo9yCrahlsDx0j6CJhO75wvGeOAKZIm2t6l5N4NwHmE\nVAofZnUTVocnKnRdXwGq9QDzdj4i6cfADbE3OJ0wH3GPpF9F+18n6Dq+FYudBlwl6UGCk36vTNUn\nA5dL+kbJM1OAmbHs2YR0DBlHAGdKmkLIeLg7icQgojNcYXm6Tg+yWSTNbfvd2IO8AjjT9hXttqsI\ng1UPMolV1E+zYhV9pQe5wmpr+ncXzaI9DcB2oz/Rdj3ItJNmVo6Ii08PERZ8rmyzPYlEF9PZaV+7\natJfIWnXeSWXP7D9qaJ12D60b61KtJpdx7QvvKmZXuC0D6tGmNRkjtmG1n5oANAhvrAsXeUgbU+l\nJ4YwkUh0OBIdsyBTjq5ykIlEYuDRwf4xOchEItE+sjCfTiU5yEQi0VY6ZUGmHGkVO5FItBVV+K9Q\n2fIpF+aXdKOkJ+Lf+XL36kq50DY9SJXoP1Z57m+S5m3MysK2zCtpvwLP1dR+LFPmNklj4+uWv5dE\nYiCRiVU0sZPmbIJGQ57DgJttr0DYNXcYzJJyYRvgZEUVr0q0TA9SgWr1F3KQtj9v+80m7CzCvEBN\nB9ks/fReEomBQwWps6JyZxVSLmwPnBNfn0PPDru2pVwYDywfe1iPSzqXEGi9pKSdJU2NPc1fA6i8\n/mMlfcZnJC0Y635UQaPxYUk3SBpRySBJy0u6SUGDcqKk5STNLenmeD5V0vbx8aOB5WLbx1R5DmCY\npAuiLZdlPWdJm0uaFJ8/U0F9qNSm7L1U0sd8RtJR0Y4HJK0j6XpJ/5RUVpBY0t7x2QdeefWVOv/Z\nEon2UkMwt9GUC4vYfjG+/g9BBQv6M+VChnr0IKfGSysAJ9tejbDH+dfAZoT4xHUl7WD7MHp6oLuo\ntz5jJhVWuh87q/sPse436S0mUcoF8dk1gQ0J0uvTgB1tr0PQWTw27uM+DPhntOf7VZ4DWCm+v1WA\nt4H9JM1B6OrvZHsNwuJXuf3mGdsQ9DHXjD3w63L3/hU/g/Gxzq8A6xN0I2fB9rhMkn6hBReq0mQi\n0ZlI5Q8wzU6cAAAgAElEQVQaT7nwMQ57qRveT92Mg8z0IB8A/kWPBtuztu+Jr9cFbrP9iu0ZBKf1\n6TJ15fUZJ8fzT5Z57mnbmQblBIJYxSxImgdYPNtDbXua7f8SfrB+FYUhbiL8eixSrooqz/3b9p3x\n9fkEPceVom3/iNfPqfA+M6YCW0r6taRNbL+Vu3d17pl7bb9j+xWCJmSav0x0HS3YaviSoqp4/Pty\nvF53yoVW6EGWU7mpRVl9xjKU6iNWHGJXYBdgIWBMVBJ/BpijzudKf43q/nWqoY+ZvceP6P1+PyKF\nZSW6kBYE+VxNUMU6Ov69Knf9Qkm/I+TFrplyodVhPvcR9BUXjHOKOwO3x3t5/cdK+owNYfsd4DlF\nAVyFHBRzEvQaX45O77NA1kapXmOl5wCWkrRBfP114A7gcWAZScvH67vl3ucsqJg+ZiLR9YjQsSp3\nFCofUi7cDawk6TlJexEc45aSngC2iOfYfhjIUi5cR4tTLtTE9ouSDgNuJXwWf7WdefNe+o8qo88I\nNK5jFZzUqZKOjPV9lTDE/4ukqYSpgceina9JulMhhOdawrzpLM9FHgf2l3Qm4YM+Jaqe7wFcGudk\n7wf+WMW2Naitj5lIdD9NJuiqkHIBwjRduefrSrmQ9CC7iMGqB9nM/8NFeyqtYCCr+fSVHuSqo9f2\n+X8pP9gas8yotutBpjmtRCLRRjpH+7EcA95BSvoDsFHJ5eNtn9UOexL10+wo5rnX32+47JIL1LUB\nbBamTW+8F9hsD/DOJ19tuOxGyy/YVNt9RZiDbLcVlRnwDtL2/u22IZFINE7RfdftYMA7yEQiMbBp\nZpGm1SQHmUgk2ofau1BWi66WO1NvxaG/tGIniqRNJV1T45m1JH0+d75dDH9KJAY12Rxkha2Gbaer\nHSS9FYdeJ8RWtoO1CLtmALB9te2j22RLItFRJAfZGdxNTrlD0vcl3S9piqSf567/REGR6A5Jf5J0\naLye13VcMG4/7IWk9STdHVV97pK0kqTZgCOBnWJvdidJ35R0UiyzjKRboh03S1oqXj9b0gmxnqck\nfaWVH04i0S46Oe3roHCQcZvj5kQhCElbEfZhrkfo3Y2R9GlJ6xIUgtYkKBTVG6T6GLCJ7bWBnwK/\nsv1hfH1x7M1eXFLmRMI+9NGEnT4n5O4tShDD2Ja4XSqR6DZU4egEun2RJlMcWhx4FLgxXt8qHpPi\n+dwEhzkPcJXtacA0SX+ps71RwDmSViCIWAyv8TzABsCX4uvzgN/k7l1p+yPgEUnlVIeIGnl7Ayy5\nVPvyQycSjZDtxW64fBjJvUMQr5lhe6yk+YGLCWpfzwBfs/1GI/V3ew8yUxxamvBvkc1BCjgq9ujW\nsr287TMq1hKYQc/nVU4BCOAXwK1xzvOLVZ4rSl7Np+z/RUkPMjGgqTD/WKfP/Gz8HmcjvrIpFxqh\n2x0kAFEL8kDgf6OYxPXAnpLmBpC0eFQSuhP4oqQ54r1tc9U8Q9CshCBiW45R9OjLfTN3vVQtKM9d\nhDwZEGTWxhd8W4lEV9CCRZpKKRfqZlA4SADbk4ApwM62bwAuBO6Oij2XAfPYvp8wTzmFoOozFcjE\nbH8L7CtpElBpn9ZvgKPiM/npi1uBVbNFmpIyBwB7RHHe3YCDmnyricQAovwCTR2LNAZukjRBPSkZ\nKqVcqN+6pObTG0lz23436kf+Hdjb9sR221WEgarmM2j3Yg8fuHux+0rNZ/RaY3z1TXeWvbfsQiOe\nBfJvclxp2gVJi9t+Po4AbyR0OK62PW/umTdsz0cDdPsiTSOMU0gPOQdhdXlAOMdEYqBSZZHm1VpO\n2Pbz8e/Lkq4gRKa8JGnRqEebT7lQN8lBlmD76+22IZEYTDS6F1vSXMAQ2+/E11sRYo4rpVyom+Qg\nEwBMn/FRU+WHD2t8OrvZvbij5iwSTdUahg9t3zR+M8Pkp15uJHVUC2huQWYR4Ir4/88w4ELb10m6\nH7gkpl94Fvhaow0kB5lIJNpGM3GQtp8ibOoovf4aFVIu1EtykIlEoq0kubNEIpGoQBLMTSQSiQp0\niC5FWQaMg5S0AGHbEMAnCHsvX4nn60VRiHbYtT/wpu0LqjyzGSEP9j1FyyQSgwGJjlHuKceAcZBx\n4nUtAElHAO/a/m3+GYXZXkWBh5YjaZjtPxR4dDNCwOs9AAXLJBKDg871jwN/q6Gk5SU9IukC4GFg\nUUnjJD0g6WFJP809+5ykI6Je4xRJK8brm0l6MG4FnBhjqpD0I0lT471fxmt3SDpO0gPAdyX9n6SD\nc/d+H+uZKmmspOWAbwHfj9c3LCmzjqR7oz2XSxqVq+toSfdFfcoN+/FjTST6jSEqf3QCA95BRlYG\njrO9aoysPyxG4K8JbBl3xmS8FPUaTwcOide+T9hSuBbwaYLU2RcJmpDr2V4TODZXx9CooPP7MrbM\nHus5CDjd9j9jW8dExZG7Sp4/Hzgk6kE+Dvwkd0+214v2/ZQySNo7/hg88Mqrr5R7JJHoWNT8XuyW\n0i0O8p+285uQd5Y0EZgIrALkHeSf498JBL04CCo+x0s6ABhpeyawBXCm7fcBbL+eq6NU9DbPn+Lz\ntwALZ4pB5YjzqnPYzjajnkNw0NVs7UWSO0skWke3OMiPtwVEsdqDgM1ir+w6eusyZhqLM4lzsLb/\njyA6OzdwT6yjUHtlKFVeaEaJYRZbE4luI/Ug+5eRBP3Ft+NG9a1rFZC0nO0pto8i9DpXIiiD7Clp\nRHxm/oLt7xSf35QwnH+PCnqQceHp/dz84m7A7QXbSSQGPKow/9gpc5Dd2CuZCDxCyA/zLGH4XItD\nJW0CfETQgrzB9oeS1gQekDQd+Au95wcrMT2meRgK7BGvXQVcKulLzJpZcTfglOiIn8yVSSQGBx3i\nDMuR9CD7EEl3AN+1Pbkd7TejB9lOsYpmefv96Q2XHTmiOaGLmR81/v0Z2sZuUrNiFastPnef6EGu\nM2asx999f9l7c88+pGYbkrYBjid0SE7v63TK3TjETiQSA4hGUy7EbKV/IESbrEpYnF21eqn66MYh\ndtuwvXG7bUgkBhpN7MVeD3gyqvog6SJCPppH+si05CC7iYkTJ7w6YrierfLIgvSWsK+HZsqmtruv\n7aUbrLcXkyZOuH7O2VRJ2HKOuCEjozTlwuLAv3PnzwGf6gu7MpKD7CJsVw2ElPRAo/NGzZRNbQ++\ntotie5tWt9EMaQ4ykUgMVJ4HlsydL0FP2uU+ITnIRCIxULkfWEHSspJmI+SXv7ovG0hD7MHFuNqP\ntKRsanvwtd1ybM+Q9F3gekKYz5m2H+7LNlIcZCKRSFQgDbETiUSiAslBJhKJRAWSg0wkEokKpEWa\nRKJDkTSS3He0RJO0r9taH/gZIQB8GEFCwrZXbFWbA4G0SNPlSFqnzOW3gGdtz6hRdkWCmnn2pQHA\n9mZ1tL8hQew3X/7cOsoPBRYpKf+vgmU3BlawfZakhYC5bT9doNzswJfL2H1kP9m9D/BzYBo9eqK2\n/ckCZb8E/BpYmODkMkc3ska5R4EfEMSZZ+ZsfqmIzd1KcpBdjqR7gHUIMm4CVifk7hkF7Gv7hipl\nHwT+yKxfmgkF2z4PWA6YnCtv2wcWLH8AoVfzEkGKLis/ukDZnwFjgZVsryhpMeBS2xsVKHsd4Uek\n9H0fW7FQH9kdyz8BbGC77m2Ckp4Evmj70TrL3Wu7T7fpdQNpiN39vADslcWHRbWTIwm9hT8DFR0k\nMMP2KU20PRZY1Y3/Ch9EcHCvNVB2R2Btgj4otl+QNItocQWWaHILXDN2A/wT+G+DZV+q1zlGbpF0\nFOH/iUzJHttTGrSjK0gOsvtZMR88a/sRSSvbfkq1NaX+Imk/4Ap6f2mKzoU9RMhh/mKdNmf8m9CT\na4QPbVuSAbJMlQW5S9Iatqc22HYzdgMcHm24l96fe5Ge9wOSLgauLCn758pFANi45C+E4f2nyzw7\naEgOsvt5WNIpwEXxfCfgkTjPVktpdvf49/u5awZqzoVFFoxt3UfvL+t2Bcs/Bdwm6a8l5X9XoOwl\nkk4F5pX0bWBP4LSC7W4MfFPS07HdbB6v0BC5SbsBTgVuAabSM0QvykhC73Or3DXTkwCuLLY3qbOd\nQUGag+xyYiqH/ejpGdwJnExYAJjT9rstbPsz5a7bLpR3J84jliv/84LltyQ4CgHX276xYLmyUl62\nq0nJ5cs3a/ekmJq434jTDz+hp8d4O/B/tt/pTzs6jeQgExWRNBzYl54vzW3AqbYL5ziQtAiwbjy9\nz/bLDdgxN0ArnXmZNtcEsl7VeNsPNlBHQ3ZL+hXwDCEPUl1TG5KWAE4EssWo8cBBtp+rUe5S4B+E\n1MMQciWtYvsr9djebSQH2eVI2gg4gllDdYqEjJwODKf3l2am7W8VbPtrwDEExyqCw/m+7csKll8d\nOA/IMkq+CnyjiCBBo+EusexBwLfpGZbuSBBrPbHVdsfy5UKRiob53AhcGNsH2BXYxfaWNcpNtr1W\nrWuDDtvp6OKDkN3xcwRHsUB2FCz7YJFr1coDC+fOF6qz/F3AZ3PnmwJ3FSz7JKEH1MhnNgWYK3c+\nFzClP+zug3/vyUWulXnmHmD93Pn6wD39YXMnH2mrYffzlu1rbb9s+7XsKFh2pqTlshNJnyQXF1iA\nIe49pH6N+ra3zmX71uzE9m0EZ1WERsNdIPQ28+9zJvUlJ23GbiR9NQtJkvRjSX+WVHRO8jVJu0oa\nGo9dCZ97LfYDzpD0pKR/Eha09i1qc7eSVrG7n1slHcOs8W0TC5T9fiz/FMFBLE19ebuvk3Q98Kd4\nvhPwtzrKPyXpJ/QeLj5VsGyj4S4AZwH3Sroinu8AnFGwXWjOboCf2L407gTagjBN8UeK5VvZkzAH\neRxh9fouCvybxf8fVpM0fzxv2bbGgUSag+xyJN1a5rJdcLtgDAdaKZ4+bvuDas+XKf9lcgsGtq+o\n9nxJ2fkIW+6yFfjxwBG23yhQ9qwyl217z4Jtr5Nv1/akIuVi2YbtjuUn2V47Bm5PtX1hq1a2Je1s\n+0+SysZY2j6hr9scSCQHmZgFSZvZviUudMxCwV7YgEPSSNtvZ72oUvqrVyXpGkJulS0J20TfJ0QA\nrFmlzA9s/0bSifTs3/4YVwgyl7Sf7ZMl/aLMbdv+aUNvoktIQ+wuRdKuts+XdEi5+64etPwZQqDy\nF8sVpUbQsaQ7bG8s6R16f1mLCif83vbBkv5C+S97zUDzBsNdLgS2JezBnsVuagTI94Xdka8B2wC/\ntf2mpEXpHaxfjmy+9YGqT81q08nx5V9t35O/FxV+BjWpB9mlSNrH9qnNBi23A0ljbE9oJtC80XCX\nZmjW7r7owUr6qu1La10rU26i7XVKrk2wPaZWm11Nu5fR09G5B0F0YSShB3U6QfhhqzrKn1fkWrX2\ni1yrULahcJf43M1FrvW13cA18e/ThEWdp3PHUwXbnljkWu7eevHf+d/Agbnjx9QR2tStRxpidymS\nqk6uu5jwwZ62j5e0NSF+cjdCj6yaAlCe1UpsGgbU0yPZHTi+5No3y1wrx2sxxCVbQd+ZGuEukuYA\n5gQWjAstWWjPSGDxgjZDg3bb3jb+XbaOtgCQ9Dng88DiJf/2I4Fqup9zEfbMDyPEqWa8A3y1Xju6\njeQgu5dMs3EjYFXg4nj+VeCRgnVkDuLzwLm2H1YBCSBJhwM/AkZIejtX14cUSCcqaWfg68CykvJ5\njucBii6UNBLusg9wMLAY4fPL3uvbwEmttlvlxY0/xtVDs14gzD9uR8+/PQRH970qdd5KCOU6y0Hh\nac5w2e/XsncwkOYguxwFwdyNHdXD4/7q8bZrTsDHUJnFgWWBNQm5h29zwXkpSUfZPrwBm5eObR4F\nHJa79Q5h2FdVCb1ZJB3ggtsKS8o1ZXeFkKwMu0BolqThrmOvfK7cOoRYz6wX+RLwLdcR3tSNJAfZ\n5Uh6nKBO/Xo8n4+whWyl6iVB0hBgLcL815tx8WAJ1yGiGttbAZgju2b773W+jcJIqhaWYtvlwlnK\n1bM6oeedt7twqoh2IWkFgoMutb3WCvyDwMGxR4mkTYHjXSW0aDCQhtjdz9HApNg7EUGZ54iCZTcg\nLGy8F+fz1qHY/B8Akr5FWABYgpB2YX3gbqBokPr6hGHyKsBshB7se64eJvRemWtzAXsR5lFrOsi4\n8r8pwcn8jbCX/Q6gkINs0O58+WZUlM4ipHs4DvgsYVqhyPbOj1yyPVJSvVqU3Ue7V4nS0fqDoOq9\nfTw+UUe5LI/NmsAkYH/g9jrKTyX0YibH85WBP9dR/gFg+dj2UMKX/ag6ys9DWI19mqjsU4fdQ4jC\nGoTkWzf2o92nExSUNovHWcDpBctOyN5D6bUa5Y4D/kDY/bMRcAJwLDAaGN3u/4fbdaQe5OBgKPAK\nYcSwoqQVXWyYO8O2JW0PnGT7DEl71dHuNNvTJCFpdtuPSao5tM9j+0lJQ23PBM6SNImQkqAicSrg\nEGAXgqNZxwW3+UXet/2RpBkKqVdfBpZstd051nXvoe0tcQhchA/i1MgTkr5L2JEzd4FyY+PfUtX0\n9RjEqReSg+xyJP2aIBLxMLkMe0ARB/lOXJHeDdgkfvGG19H8c5LmJQhG3CjpDaCQKnfkv5JmAyZL\n+g0ht03V4WIU5vgSYbV8DTcmsvtAtPs0worwu4SpgZbZXcJMScvZ/ifUraJ0ECFU6UDCdMJm9KTO\nqIhTyoWypEWaLicu0ox2nSITsewnCGEr99seL2kpYFM3sFgRd5eMAq51wVXWuCr8MsEpfy+WP9n2\nk1XKfERQ75lBA9scy9S3DDDS9S1M1W13SfnNCcPqXipKzs0R9hVJrKI6yUF2OZKuBb7aYE8q+7Kv\nYPumGCM31AXzlEg6AzjR9uTctSNsH9GILa2myTjEvralLhWlSvu/M1xhH3gNsQps/6SgyV1JcpBd\njqTLCYssN1NnClGFbIB7A/PbXi6GkPzR9uYF236OsHvld7bPiddm2fNbptxUqn/Zi2YXrItm4xD7\nyu64oydLtGaC0MYfbU+rUqbs/u9c2xX3gUsaCuw/2HuL5UgOssuRVHb+KXNYNcpOJkzS3+uoRShp\nqu01CrY9kRBqcj7wL8L82P2uoWuoClkFc7bXM4/Zb/SV3ZIuIQSXnx8vfR2Y13bLtv5Jus/2eq2q\nf6CSFmm6nCKOsAof2P4w210Y91LX84sq228BX5R0BCGeb1StQu12gJImEHaV/Kme1e8+tHt126vm\nzm+VVGh7qELCr3JSa7USft0h6feELakfx5LWM/fajSQH2aVIusT21yoN+woO926XlO2p3pIw7PtL\nHWZ8vB/Z9hHR8VTcF1yKeutJzkZY9CgccN0EOxFiF++X9ABhweQGFxxu9YHdEyWt76jPKOlTFNd5\nHJt7PQdh731Z+bQSstS8+W2kgza8JyMNsbsUSYvafrHSsK9IbyeG9ewFbEVYTb2eELDc7//TKHRj\ntydk3jus1vN91OYQgoDuKYQwm7MI2+8KK4vXY3fux2w4YYHmX/F8aeCxkl5lPe8j6To2SHKQgwhJ\nCwKvFXFwceL+XNu7NNBOU4riNepuSW6WMu2MJvQiP0/4YbiAsGiymxvIFV3E7r6YwyxZiR9C6FHu\n6xp7quMq9rG234zn8xH2ZpcVXB4spCF2lxL3Ax9NkNn6BUHHcUFgiKRv2L6uWnnbMyUtLWk22x/W\n07btjePfeRqzPqDeOXGyL3vFldy+Ik4FvEmYhzwsF2Jzr6SNKpf8uHxDdpc6QEkLkxOcKMixudcz\ngGcIKRxqsW0+pMf2G5K+SNjXPWhJDrJ7OYmgyTiKkF/mc7bvkbQyQUS2qoOMPAXcqaBtmJ+4r5bP\nJtvqV5E6hqj5nDjZl337gmUbIg6rL7f9q3L3bZdNZFZCU3ZL2o7g6BYjBJwvTcg5s1q1ctG+zxZt\np4Sh+R/DGGo0W4N1dQ3JQXYvw2zfACDpyGzCP+6HLlrHP+MxhCD8UJQs6VW5hmomv/r4QbueHNx9\nQtyD/SWgrIMsWEezdv+CoHx0k0P6188ScurUROWTtL1FEKyYXOZexkWE7aBnxvM9CdMKg5rkILuX\nvFRVqTp0oYlnx8ReUbDBRXfQuIGUAeVQ+bQRbwEP2L6qL9qowE2SDmXWkJdCPd8+sHu67dckDZE0\nxPatMQSnCGPjkUUbbEtQZfqOpEtt/6ZcIdu/kjQF2CJe+o3tvxZss2tJizRdiqSZhC+3gBHAf7Nb\nwBy2a4pOSBpLWLnNeo9vEfLUTKhcapY6GhbMlTSOIJGWZeT7MkG6bAGCiO/BRe2ohxhLWIoLxBJm\n5ZuyW9JNwA4E4dsFCcPsdW1vWKDtvwOfz7aWSpob+CshjeyESivhkkYQ4l4/krQ8sCIhtKml6u0d\njztAcy0dnXkQeh6b5M43po5Md8C3CNqKbwC3Enqyt9RR/h7C3u/sfBhBVWco8Ei7P59W2U0Q+B0S\ny+1OUOaZv2DbjwHDc+ezE0KEACZVKfcAQQVoUcKc6RWEKIa2f57tPOqRYEoMPmbaHp+d2L6D6hny\nSjmIEID8rMPiwdqE1eGizEdvLcO5CI5iJrl95X2NpOGSDpR0WTy+q6DyXZRm7f6p7Y9sz7B9jsMe\n6R8WbPsCwmr7zxSU0e8ELpQ0F9WTtQ2x/V9Cb/ePtndkVm3IQUeag0xU43ZJpxJWvU3YYXJbFmvn\n2uo2zQrm/oagqXgbPekifhW/7DfV+2bq4BRCsPbJ8Xy3eO1bBcs3a/eWzOoQP1fm2izY/oWCglMW\njvQd29kunGoxrUMkrRuf+Xa8NrSArV1NmoNMVKQP1G2uIARbH0wQbn2DMPz7fB02LEoQzIAgdPFC\n0bKNIulBlwRWl7tWo4667Za0L2E75ycJ0QMZ8wB32i66kr0xQaLuLEkLAXPbLjevmi+zGXBobOeX\nCiK9h9rer0ib3UpykIl+QT2Cude5YOB53Ka3C/BJ20cqCPZ+wvZ9LTQ1UyH6qnsrel/mGjJtufIN\n2S1pFGF4PkvaWBdfQf8ZYRV7JdsrSloMuNR2zQD3xKwkB5moiKTzgO86KPJkW+HOdHE9yPWBhx3D\ng2K40Cq27y1Y/hRCuNJmtleJK+I32F63RtGmUJOK3n1ld+lOGtv/KlBmMmGud6J7JOqmuIY4SVy5\nPgRYhtzUm+2t6rG520hzkIlq3EGY8D8EWBz4PvC/dZQ/hZAqNuPdMteq8Snb6ygkvMJh+1vLd3fY\nvllBHLiwoncJTdkdt/j9jgZ20gAf2rYkx7rmKtjsZYStledTPP9N15McZKIitk+V9DAhROdVYG3b\n/6mjCjk3RHGIsavn/7npUTQj+7IvRO8A+FYyhp7e1FqScPFcPM3a/X80uJMGuCQurM2roAi/JyGN\nbC0+sn1iHTYOClKYT6IiknYDzgS+AZwN/E1S4YUK4KkYLjM8HgcRhq1FOYEQj7ewpF8SerQNbwEs\nSpxa+C0h7nPdeIytWqg3zdo93fZrhJXlIXFoX6h9278l9AYvJ/SAf+piqRSukrS3pIUkjcyOOmzu\nStIcZKIikq4E9rb9cjxfDxjngnJfcQ7tBMIKNoQQl4Oz+grWsTKwOWEu8Gbbj9bxFhpC0qPAqm7i\ny9GM3c3spClT1xBgZ9tV91VL+neZy7a9VL1tdhPJQSbqQg3InzXYzlDCAs/KrW6rTNuXAgfafrGB\nsk3bHecN3yeM8HYhrP5fEHuVlcqMBPYnzBVfDdwYzw8FHrTdUhWkbiXNQSYqImlFwqLKIrZXVxCR\n3Y4wR1ak/BLAifQELY8HDrL9XK2yDnqUj0taqsjqbV+gntSp8wCPSLqP3pkgy6ZOzdNHdi8MvOiQ\nxfCcuE96EUKGyEqcR4gzvZsQ0P4jQu91B1dX8fmY2Otdld4r5xc29A66hNSDTFRE0u2EletTcyEj\nD9levWD5G4ELCV9eCAsNu9jesmD5vxNCVu6jt6pOTUfVCGoidWpJPU3ZrZAHZ0P3aDPORgjgrhgm\npFy2ydiLfRFYylVSxZaU/zEhtcbKBAX1rYE7XEz/smtJPchENea0fZ9660fWsxd7Idtn5c7PllSP\nAk+/Jq3PO0BJnyDshDFhJ0w9q/fN2j0sP43hkFmyVpjQ9NzzMyU9V9Q5RnYC1iLET+4WdwKdXY/R\n3UhykIlqvCppOXrCVb5C6JkU5TVJuxL2cgPsTPVhYi9q9dgk3W17gzrsKYSkbwE/JSixCzhRQXT4\nzOolA31g9yuStrN9dXx+e0KYVTXWlPR21gQhE+XbFM8D9H50rDMkzQP8hxB/OahJDjJRjf2BccDK\nkp4naBrWk8RrT8Ic5HEEJ3sX8M0+tK/efC1F+T4h5vM1AEkLEGwv5CALUMvu7wAXSDopnj9HEMyo\niO1mhSUmSZqX8B4fAN4mTBEMatIcZKImcVV1iEsUxSXtbvucOus62HZRdexadU0suj+6znrvAjYt\nmQO8rZEwmwr1F7JbQewWR/Hb3PW6P/ca7YiwV/zFeL48MNK11Zq6nuQgEw3TiIOS9K++iq1roYM8\nF1gDuIrQ892eIB48BWonLStQf1N2t+J917P4NphIQ+xEMxTO/tVkmf6oK0+WrCwjyyPTVBrbHM3a\n3Yr3PVnS2rYntaDuAUtykIlmaGT4UVeZqCC0gu2bYjzgsNxQv+q8XKM4JiuL7Q8h6Cm+XaXILLTY\n7j4b9kka5pB3Zm3gfkn/pCeXkVvRQx9IJAeZaIayPRlJ71D+S5wlECtWeRBb2BuYH1gOWAL4I2EL\nH7YfqtPeou1eSFgomQncD4yUdLztYzrE7r7sQd5HUFdqSWzpQCc5yEQz3Fnuou2+GoruT4hFvDfW\n+0Tc391qVrX9tqRdgGsJ4rUTgEIOktbbXfZzbxABOIoDJ3qTHGSiIpIWIajQLGb7c5JWBTawfQaA\n7e+22IQPYpB0Zs8w+nB4WYXhCkm6dgBOsj0901csSFN2S5qdkDxrGXqL1x4Z//bl576Qgt5nWZpd\nkMVAf3UAAAelSURBVBroJLmzRDXOJmw7Wyye/4OQX6a/uF3SjwhBz1sS8kz/pR/aPZWQ+nQu4O9x\nPrGeOchm7b6KsHI+gzAfmB2tYCghA+M8FY5BTQrzSVRE0v2215U0KbcXe3JRubM+aH8IsBdhj7AI\nzvr0ZmTImrAlW8wo8mxTdvdnyE2rQqW6hTTETlTjvbiLJNtquD7wVj+2vwMhef1p/dhmljzrZ4R0\nrQC3A0dS/L03a/ddktawPbXB8vXQqlCpriANsRPVOISgLbicpDuBc4ED+rH9LwL/kHSepG1VX7qG\nZjgTeAf4WjzeJiTxKkqzdm8MTIiyaVMkTZU0pc46ilIoAdtgJQ2xE1WJX+6VCD2Nx21Pr1Gkr9sf\nDnyOoDazMXCj7W+1uM1ZphHqnVpoxu445zkLtp8t2n6ib0hD7EQt1qNnNXUd1Ze8qmniCvK1hGH+\nCMLwtaUOEnhf0sa27wCQtBFB4bswjdgtaWQMSH+n2nOJ/iP1IBMVUUhetRwwmZ5UoLZ9YD+1n/XA\nNgVuAy4h5JeuR5OykXbXJEwnjCL0nF8Hvmn7wYLlG7Jb0jW2t5X0NMGx5ucHbfuTdb6VRJMkB5mo\niPogeVWT7f8JuBi41vXlpe6r9kcCNLDNsK12J/qO5CATFVETyasGMrUCtfuhfRF0N5e1/QtJSxHk\nyAa9PmN/k+YgE9VYkAaTVzWDpDtsb1xmT3dRdexmuYoQ0jOB3PuuRR/afTLwESFd7i8Ic5KXE/Jz\nJ/qR1INMVEQVkljVSikw0Gm3NmIWvF0SoP+g7TXbZdNgJcVBJioSHeEzwPD4+n6g31Sm4yJRzWst\n4C5JazRauA/snq6QmTAL0F+I0KNM9DNpiJ2oSBnZrsXJyXb1A6uV2DMMGNOqxiRNJTilYcAekp4i\nDLGzIfLoglU1a/cJwBXAwpJ+CXyFfs7wmAgkB5moRlvkxiQdTkh8n2Xmg+CkPiQkEWsV2zZTuK/s\ntn2BpAmEHyIBO9h+tBnbEo2R5iATFZF0r+1PZXNhsSc0sY6eVLPtH2X78P5oK7Y3B0Eod3lgKnBG\nIzGXzdot6Tzbu9W6lmg9qQeZqEapbNd+9I/cGAC2D5c0H7ACuVSptv/eoibPAaYD4wnbBFcFDipa\nWNLKth8DLpU0i0JOHVkCS4foQ2nh1EKiMqkHmahIiWwXwPW2T+/H9r9FcFBLEHbzrA/cbXuzFrU3\n1fYa8fUw4L56pMAkjbO9t6Rby9x2LbvzQ3Tgv/TspPkQGNefvelEIDnIxCxI2h5YwvYf4vl9wEKE\nBYwf2L6sn+yYSoj9u8f2WpJWBn5l+0staq+XNmK7tBL7e2ohUZkU5pMoxw8IMmcZsxGGeJsC+/aj\nHdNsT4OwuyUOX1dqYXtrSno7Hu8Ao7PXuUWXmkj6qqR54usfS/qzpLWLlo9TC9tJ+m08mlo8SjRO\nmoNMlGM22//Ond9h+3XgdUlz9aMdz0maF7gSuFHSG0DLJL9sD+2jqn5i+1JJGwNbEJJ9/RH4VJHC\nko4iRA9cEC8dJGlD2z/qI/sSBUlD7MQsSHrS9vIV7v3T9nJtsOkzBHWd62x/2N/t10Nu1f8oYKrt\nC/O7YgqUnwKsZfujeD4UmNRf0QOJHtIQO1GOe2OQeC8k7UPIo9wvSJo/OwhhN3fQP1kNm+V5SacS\nJM/+FsUv6v2uzZt7ParPLEvURepBJmYhBoNfSdhFkoWmjAFmJwQtv9RPdjwDLAm8QVjRnRf4D/AS\n8G3bE/rDjnqRNCewDaH3+ISkRYE1bN9QsPzOwNHArYT3/WngMNsXt8rmRHmSg0xURNJm9MTkPWz7\nln5u/zTgMtvXx/OtCDJkZwHH2y40p9cOoujuJvF0fB1iuyKENc2gR73nPtv/6XsrE7VIDjLRseTj\nEnPXptgerX5MP1svkg4Cvg38OV7akRDHeGLB8rO870R7SKvYiU7mRUk/BC6K5zsBL8VFi05Wt9kL\n+JTt9wAk/Rq4GyjkIIGJkta1fX+rDEwUIznIRCfzdUJ+6isJizN3xmtDCelYOxXRk8OH+Lqe/NOf\nAnaNc7DvUb+aUKKPSA4y0bHYfhU4QNJcWW8sx5PtsKkgZxEiAa6I5zsAZ9RRfuu+NynRCGkOMtGx\nSNoQOB2Y2/ZSceFjH9v7tdm0mkSxio3j6XjbkwqU6RM1oUTfkRxkomORdC9BLPbqXOqBtqZDqEaz\nDk7SxfRWE3rWdmE1oUTfk4bYiY7G9r9D5MvHzKz0bAdQKpe2CnBwHeVXzakJnUE/BuUnypMcZKKT\n+XccZlvScIL0WScrazfr4KZnL2zPKPlhSLSB5CATncx3gOMJuXCeB24gpIHoVJp1cGuWpGrIUjf0\nV7rbRAlpDjKR6CMkzSSE5UB0cPQI3yYHNwBJDjLRcUj6aZXbtv2LfjMmMahJDjLRcUj63zKX5yLs\nUFnA9tz9bFJikJIcZKKjicrcBxGc4yXAsbZfbq9VicFCWqRJdCRRA/IQYBdC+Mw6tt9or1WJwUZy\nkImOQ9IxwJeAcQQdxXfbbFJikJKG2ImOQ9JHBLHeGfRWEE+rwYl+JTnIRCKRqEDKSZNIJBIVSA4y\nkUgkKvD/26kDAQAAAABB/tYTbFAQCRJgCBJgBHb6qR/d0Gn3AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7fb01a54add8>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import statnlpbook.util as util\n",
    "util.plot_confusion_matrix_dict(cm_dev,90, outside_label=\"None\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The confusion matrix can give you hints on what type of errors you should look for and improve upon. This macro view on your model's performance is often more powerful when combined with a micro view on the instances that produce these errors. You can find errors of a specific type using `bio.find_errors` as shown below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(<statnlpbook.bio.EventCandidate at 0x7fbd5d922470>, 'Binding', 'None'),\n",
       " (<statnlpbook.bio.EventCandidate at 0x7fbd5d6f6198>, 'Binding', 'None'),\n",
       " (<statnlpbook.bio.EventCandidate at 0x7fbd5d44d668>, 'Binding', 'None')]"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "errors = bio.find_errors(\"Binding\",\"None\", event_dev, event_dev_guess)[:3]\n",
    "errors"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "These errors you can then inspect in detail via `show_event_error`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Gold</th>\n",
       "      <th>Guess</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>Binding</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>The <font color='red'>[</font>additional<font color='red'>]</font> <font color='red'>[</font>function<font color='red'>]</font> <font color='red'>[</font>that<font color='red'>]</font> is <font color='red'>[</font>required<font color='red'>]</font> <font color='red'>[</font>for<font color='red'>]</font> <font color='red'>[</font>initiation<font color='red'>]</font> <font color='red'>[</font>of<font color='red'>]</font> the lytic viral life cycle is likely <font color='red'>[</font>to<font color='red'>]</font> <font color='red'>[</font>require<font color='red'>]</font> <font color='red'>[</font>phosphorylation<font color='red'>]</font> <font color='red'>[</font>of<font color='red'>]</font> serine 186 <font color='red'>[</font>of<font color='red'>]</font> the <font color='red'>[</font><font color='blue'>[ZEBRA]</font><font color='red'>]</font> <font color='red'>[</font>protein<font color='red'>]</font> , which may <font color='red'>[</font>influence<font color='red'>]</font> either DNA <font color='green'>recognition</font> or <font color='red'>[</font>transcriptional<font color='red'>]</font> <font color='red'>[</font>activation<font color='red'>]</font> <font color='red'>[</font>of<font color='red'>]</font> lytic viral <font color='red'>[</font>promoters<font color='red'>]</font> in a chromatinized viral episome .\n",
       "    <div id='displacy31' style=\"overflow: scroll; width: 5000px;\"></div>\n",
       "    <script>\n",
       "    $(function() {\n",
       "    requirejs.config({\n",
       "        paths: {\n",
       "            'displaCy': ['/files/node_modules/displacy/displacy'],\n",
       "                                                  // strip .js ^, require adds it back\n",
       "        },\n",
       "    });\n",
       "    require(['displaCy'], function() {\n",
       "        console.log(\"Loaded :)\");\n",
       "        const displacy = new displaCy('http://localhost:8000', {\n",
       "            container: '#displacy31',\n",
       "            format: 'spacy',\n",
       "            distance: 150,\n",
       "            offsetX: 0,\n",
       "            wordSpacing: 20,\n",
       "            arrowSpacing: 3,\n",
       "\n",
       "        });\n",
       "        const parse = {\n",
       "            arcs: [{\"dir\": \"left\", \"start\": 0, \"label\": \"det\", \"end\": 2}, {\"dir\": \"left\", \"start\": 1, \"label\": \"amod\", \"end\": 2}, {\"dir\": \"left\", \"start\": 2, \"label\": \"nsubjpass\", \"end\": 5}, {\"dir\": \"left\", \"start\": 2, \"label\": \"nsubj\", \"end\": 15}, {\"dir\": \"left\", \"start\": 4, \"label\": \"auxpass\", \"end\": 5}, {\"dir\": \"right\", \"start\": 2, \"label\": \"rcmod\", \"end\": 5}, {\"dir\": \"right\", \"start\": 5, \"label\": \"prep_for\", \"end\": 7}, {\"dir\": \"left\", \"start\": 9, \"label\": \"det\", \"end\": 13}, {\"dir\": \"left\", \"start\": 10, \"label\": \"amod\", \"end\": 13}, {\"dir\": \"left\", \"start\": 11, \"label\": \"amod\", \"end\": 13}, {\"dir\": \"left\", \"start\": 12, \"label\": \"nn\", \"end\": 13}, {\"dir\": \"right\", \"start\": 7, \"label\": \"prep_of\", \"end\": 13}, {\"dir\": \"left\", \"start\": 14, \"label\": \"cop\", \"end\": 15}, {\"dir\": \"left\", \"start\": 16, \"label\": \"aux\", \"end\": 17}, {\"dir\": \"right\", \"start\": 15, \"label\": \"xcomp\", \"end\": 17}, {\"dir\": \"right\", \"start\": 17, \"label\": \"dobj\", \"end\": 18}, {\"dir\": \"left\", \"start\": 18, \"label\": \"nsubj\", \"end\": 29}, {\"dir\": \"left\", \"start\": 20, \"label\": \"nn\", \"end\": 21}, {\"dir\": \"right\", \"start\": 18, \"label\": \"prep_of\", \"end\": 21}, {\"dir\": \"left\", \"start\": 23, \"label\": \"det\", \"end\": 25}, {\"dir\": \"left\", \"start\": 24, \"label\": \"nn\", \"end\": 25}, {\"dir\": \"right\", \"start\": 21, \"label\": \"prep_of\", \"end\": 25}, {\"dir\": \"left\", \"start\": 28, \"label\": \"aux\", \"end\": 29}, {\"dir\": \"right\", \"start\": 18, \"label\": \"rcmod\", \"end\": 29}, {\"dir\": \"left\", \"start\": 30, \"label\": \"preconj\", \"end\": 32}, {\"dir\": \"left\", \"start\": 31, \"label\": \"nn\", \"end\": 32}, {\"dir\": \"right\", \"start\": 29, \"label\": \"dobj\", \"end\": 32}, {\"dir\": \"left\", \"start\": 34, \"label\": \"amod\", \"end\": 35}, {\"dir\": \"right\", \"start\": 29, \"label\": \"dobj\", \"end\": 35}, {\"dir\": \"right\", \"start\": 32, \"label\": \"conj_or\", \"end\": 35}, {\"dir\": \"left\", \"start\": 37, \"label\": \"amod\", \"end\": 39}, {\"dir\": \"left\", \"start\": 38, \"label\": \"amod\", \"end\": 39}, {\"dir\": \"right\", \"start\": 35, \"label\": \"prep_of\", \"end\": 39}, {\"dir\": \"left\", \"start\": 41, \"label\": \"det\", \"end\": 44}, {\"dir\": \"left\", \"start\": 42, \"label\": \"amod\", \"end\": 44}, {\"dir\": \"left\", \"start\": 43, \"label\": \"amod\", \"end\": 44}, {\"dir\": \"right\", \"start\": 29, \"label\": \"prep_in\", \"end\": 44}],\n",
       "            words: [{\"text\": \"The\", \"tag\": \"DT\"}, {\"text\": \"additional\", \"tag\": \"JJ\"}, {\"text\": \"function\", \"tag\": \"NN\"}, {\"text\": \"that\", \"tag\": \"WDT\"}, {\"text\": \"is\", \"tag\": \"VBZ\"}, {\"text\": \"required\", \"tag\": \"VBN\"}, {\"text\": \"for\", \"tag\": \"IN\"}, {\"text\": \"initiation\", \"tag\": \"NN\"}, {\"text\": \"of\", \"tag\": \"IN\"}, {\"text\": \"the\", \"tag\": \"DT\"}, {\"text\": \"lytic\", \"tag\": \"JJ\"}, {\"text\": \"viral\", \"tag\": \"JJ\"}, {\"text\": \"life\", \"tag\": \"NN\"}, {\"text\": \"cycle\", \"tag\": \"NN\"}, {\"text\": \"is\", \"tag\": \"VBZ\"}, {\"text\": \"likely\", \"tag\": \"JJ\"}, {\"text\": \"to\", \"tag\": \"TO\"}, {\"text\": \"require\", \"tag\": \"VB\"}, {\"text\": \"phosphorylation\", \"tag\": \"NN\"}, {\"text\": \"of\", \"tag\": \"IN\"}, {\"text\": \"serine\", \"tag\": \"NN\"}, {\"text\": \"186\", \"tag\": \"NN\"}, {\"text\": \"of\", \"tag\": \"IN\"}, {\"text\": \"the\", \"tag\": \"DT\"}, {\"text\": \"ZEBRA\", \"tag\": \"NN\"}, {\"text\": \"protein\", \"tag\": \"NN\"}, {\"text\": \",\", \"tag\": \",\"}, {\"text\": \"which\", \"tag\": \"WDT\"}, {\"text\": \"may\", \"tag\": \"MD\"}, {\"text\": \"influence\", \"tag\": \"VB\"}, {\"text\": \"either\", \"tag\": \"CC\"}, {\"text\": \"DNA\", \"tag\": \"NN\"}, {\"text\": \"recognition\", \"tag\": \"NN\"}, {\"text\": \"or\", \"tag\": \"CC\"}, {\"text\": \"transcriptional\", \"tag\": \"JJ\"}, {\"text\": \"activation\", \"tag\": \"NN\"}, {\"text\": \"of\", \"tag\": \"IN\"}, {\"text\": \"lytic\", \"tag\": \"JJ\"}, {\"text\": \"viral\", \"tag\": \"JJ\"}, {\"text\": \"promoters\", \"tag\": \"NNS\"}, {\"text\": \"in\", \"tag\": \"IN\"}, {\"text\": \"a\", \"tag\": \"DT\"}, {\"text\": \"chromatinized\", \"tag\": \"JJ\"}, {\"text\": \"viral\", \"tag\": \"JJ\"}, {\"text\": \"episome\", \"tag\": \"NN\"}, {\"text\": \".\", \"tag\": \".\"}]\n",
       "        };\n",
       "\n",
       "        displacy.render(parse, {\n",
       "            uniqueId: 'render_displacy31'\n",
       "            //color: '#ff0000'\n",
       "        });\n",
       "        return {};\n",
       "    });\n",
       "    });\n",
       "    </script>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bio.show_event_error(*errors[1])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It can also be very useful to inspect your feature map for the given instance. Sometimes this leads you to find out that you have a bug in your feature calculation, or that the feature representation is still insufficient for other reasons."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "defaultdict(float,\n",
       "            {'Number of Proteins=1': 1.0,\n",
       "             'Parents: nn->activity': 1.0,\n",
       "             'Protein = EGR-1': 1.0,\n",
       "             'Protein length = 1': 1.0,\n",
       "             'trigger_word=binding': 1.0})"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "event_feat(errors[0][0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <font color='red'>Assessment 2</font>: Assess Accuracy (50+10 pts) \n",
    "\n",
    "We assess how well your model performs on some unseen test set. We will look at the F1 across all event types, and will score them as follows:\n",
    "\n",
    "* 0-40pts: 17% <= F1 < 60%, linear\n",
    "* 40-50pts: 60% <= F1 < 70%, linear\n",
    "* 50-60pts: 70% <= F1 < 80%, linear\n",
    "\n",
    "The **linear** mapping maps any F1 value between the lower and upper bound linearly to a score. For example, if your model's F1 score is $F=55$, then your score is $40\\frac{F-17}{60-17}$. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.6677206851119895"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#! ASSESSMENT 2 - DO NOT CHANGE, MOVE NOR COPY\n",
    "_snlp_event_test = event_dev # This line will be changed by us after submission to point to a test set.\n",
    "_snlp_event_test_guess = predict_event_labels([x for x,_ in _snlp_event_test[:]])\n",
    "_snlp_cm_test = bio.create_confusion_matrix(_snlp_event_test,_snlp_event_test_guess)  \n",
    "bio.evaluate(_snlp_cm_test)[2] # This is the F1 score"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <font color='orange'>Mark</font>:  Your solution to Task 2 is marked with ** __ points**. \n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <font color='blue'>Task 3</font>: Describe your Approach\n",
    "\n",
    "Enter a 500 words max description of your approach **in this cell**. Also provide an **error analysis** of the types of errors your system still makes, with suggestions on how to improve it further. Should you need to include figures in your report, make sure they are Python-generated. For that, feel free to create new cells after this cell (before Assessment 3 cell). Link online images at your risk."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <font color='red'>Assessment 3</font>: Assess Description (30 pts) \n",
    "\n",
    "We will mark the description along the following dimensions: \n",
    "\n",
    "* Clarity (10pts: very clear, 0pts: we can't figure out what you did)\n",
    "* Creativity (10pts: we could not have come up with this, 0pts: Use only word based features of the trigger word)\n",
    "* Substance (10pts: implemented complex state-of-the-art classifier, 0pts: Only use what is already there)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Overfitting is a common problem in machine learning, where a model performs well on training data but does not generalize well to unseen data (test data). Overfitting occurs when a model is excessively complex, such as having too many parameters relative to the number of observations. A model that has been overfit has poor predictive performance, as it overreacts to minor fluctuations in the training data.\n",
    "\n",
    "On the contrary, our model can also suffer from underfitting (high bias), which means that our model is not complex enough to capture the pattern in the training data well and therefore also suffers from low performance on unseen data.\n",
    "\n",
    "Regularization is a way of finding a good bias-variance tradeoff by tuning the complexity of the model. It is a very useful method to handle collinearity (high correlation among features), filter out noise from data, and eventually prevent overfitting.\n",
    "\n",
    "L2 regularization\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Training dataset data structure\n",
    "\n",
    "\n",
    "## Feature extracted\n",
    " - \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <font color='orange'>Mark</font>:  Your solution to Task 3 is marked with ** __ points**.\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <font color='orange'>Final mark</font>: Your solution to Assignment 2 is marked with ** __points**. "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
